# CE_MethylRAD_analysis_2018
carryover effects DNA methylation analysis for eastern oysters (Crassostrea virginica) using MethylRAD - samples and sequences from Sarah Gignoux-Wolfsohn in 2018

**Reference Genome files used:**
NCBI RefSeq assembly: GCF_002022765.2 https://www.ncbi.nlm.nih.gov/genome/annotation_euk/Crassostrea_virginica/100/

GTF file: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_002022765.2/

## [analysis folder](https://github.com/jgmcdonough/CE18_methylRAD_analysis/tree/master/analysis)
contains scripts and generated data files/figures from the various analyses
- **[Differential Methylation Analysis](https://github.com/jgmcdonough/CE18_methylRAD_analysis/tree/master/analysis/diff_methylation_analysis)** - identified significantly differentially methylated genes between treatments
- **[GO enrichment analysis](https://github.com/jgmcdonough/CE18_methylRAD_analysis/tree/master/analysis/GO_enrichment_analysis)** - matching gene ontology (GO) terms with significant differentially methylated genes, and testing for enrichment of GO terms
- **[KEGG pathway analysis](https://github.com/jgmcdonough/CE18_methylRAD_analysis/tree/master/analysis/KEGG_pathway)** - aligning identified significant genes to pathways with KEGG

## [pipeline counts folder](https://github.com/jgmcdonough/CE18_methylRAD_analysis/tree/master/pipeline_counts)
contains scripts and files of sequences counts per sample for each step in quality pipeline

## [quality check folder](https://github.com/jgmcdonough/CE18_methylRAD_analysis/tree/master/quality_check)
contains scripts and files used to quality check sequences in prep for analysis








