{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8b0f0cb1-ee02-4510-b59f-65856356a6d8",
   "metadata": {},
   "source": [
    "## GSEA analysis\n",
    "\n",
    "following pipeline [here](https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html)\n",
    "\n",
    "#### creating ranks\n",
    "\"Rank all genes based on their fold change. We need to exclude genes for which we do not have Entrez IDs. Also, we should use the shrunk LFC values.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8ff3d6f8-4e74-4807-9c94-1b85dcd58a57",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#BiocManager::install('fgsea')\n",
    "library(fgsea)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "616c324e-dba7-451e-b5b1-8f7567a9ae19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>logFC</th><th scope=col>pval</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>LOC111099115</td><td>1.7885864</td><td>4.161092e-06</td><td>0.0002004183</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>LOC111099216</td><td>1.4828614</td><td>7.544561e-04</td><td>0.0125496496</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>LOC111099417</td><td>2.7487429</td><td>6.452048e-05</td><td>0.0019567084</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>LOC111099424</td><td>3.4384445</td><td>3.211770e-05</td><td>0.0010745946</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>LOC111099722</td><td>0.8993515</td><td>1.093514e-04</td><td>0.0028699825</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>LOC111099918</td><td>0.6567512</td><td>3.539745e-03</td><td>0.0407210025</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & gene & logFC & pval & padj\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & LOC111099115 & 1.7885864 & 4.161092e-06 & 0.0002004183\\\\\n",
       "\t2 & LOC111099216 & 1.4828614 & 7.544561e-04 & 0.0125496496\\\\\n",
       "\t3 & LOC111099417 & 2.7487429 & 6.452048e-05 & 0.0019567084\\\\\n",
       "\t4 & LOC111099424 & 3.4384445 & 3.211770e-05 & 0.0010745946\\\\\n",
       "\t5 & LOC111099722 & 0.8993515 & 1.093514e-04 & 0.0028699825\\\\\n",
       "\t6 & LOC111099918 & 0.6567512 & 3.539745e-03 & 0.0407210025\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | logFC &lt;dbl&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | LOC111099115 | 1.7885864 | 4.161092e-06 | 0.0002004183 |\n",
       "| 2 | LOC111099216 | 1.4828614 | 7.544561e-04 | 0.0125496496 |\n",
       "| 3 | LOC111099417 | 2.7487429 | 6.452048e-05 | 0.0019567084 |\n",
       "| 4 | LOC111099424 | 3.4384445 | 3.211770e-05 | 0.0010745946 |\n",
       "| 5 | LOC111099722 | 0.8993515 | 1.093514e-04 | 0.0028699825 |\n",
       "| 6 | LOC111099918 | 0.6567512 | 3.539745e-03 | 0.0407210025 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene         logFC     pval         padj        \n",
       "1 LOC111099115 1.7885864 4.161092e-06 0.0002004183\n",
       "2 LOC111099216 1.4828614 7.544561e-04 0.0125496496\n",
       "3 LOC111099417 2.7487429 6.452048e-05 0.0019567084\n",
       "4 LOC111099424 3.4384445 3.211770e-05 0.0010745946\n",
       "5 LOC111099722 0.8993515 1.093514e-04 0.0028699825\n",
       "6 LOC111099918 0.6567512 3.539745e-03 0.0407210025"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>405</li><li>4</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 405\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 405\n",
       "2. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 405   4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading in data file\n",
    "genes <- read.csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/significant_genes/sig_p1_wc_genes.csv')\n",
    "#head(genes)\n",
    "\n",
    "# only grabbing what I need and changing col names\n",
    "genes <- select(genes, Row.names, log2FoldChange, pvalue, padj)\n",
    "colnames(genes) = c('gene', 'logFC', 'pval', 'padj')\n",
    "\n",
    "head(genes)\n",
    "dim(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "54c4cc5a-bfe6-445a-9e1e-58386ad42a1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>entrez_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>LOC111134684</td><td>111134684</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>LOC111110475</td><td>111110475</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>LOC111132265</td><td>111132265</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>LOC111135536</td><td>111135536</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>LOC111099115</td><td>111099115</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>LOC111136502</td><td>111136502</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & gene & entrez\\_id\\\\\n",
       "  & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & LOC111134684 & 111134684\\\\\n",
       "\t2 & LOC111110475 & 111110475\\\\\n",
       "\t3 & LOC111132265 & 111132265\\\\\n",
       "\t4 & LOC111135536 & 111135536\\\\\n",
       "\t5 & LOC111099115 & 111099115\\\\\n",
       "\t6 & LOC111136502 & 111136502\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | entrez_id &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 1 | LOC111134684 | 111134684 |\n",
       "| 2 | LOC111110475 | 111110475 |\n",
       "| 3 | LOC111132265 | 111132265 |\n",
       "| 4 | LOC111135536 | 111135536 |\n",
       "| 5 | LOC111099115 | 111099115 |\n",
       "| 6 | LOC111136502 | 111136502 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene         entrez_id\n",
       "1 LOC111134684 111134684\n",
       "2 LOC111110475 111110475\n",
       "3 LOC111132265 111132265\n",
       "4 LOC111135536 111135536\n",
       "5 LOC111099115 111099115\n",
       "6 LOC111136502 111136502"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# importing file with conversion IDs\n",
    "convert <- read.csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/KEGG_pathway/p1_wc_DAVID_conversion.txt', sep='\\t')\n",
    "\n",
    "# selecting only the columns I need\n",
    "convert <- select(convert, From, To)\n",
    "\n",
    "# renaming col names\n",
    "colnames(convert) = c('gene', 'entrez_id')\n",
    "head(convert)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "56266060-56ad-45ad-ab25-10ddfcb903ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>entrez_id</th><th scope=col>logFC</th><th scope=col>pval</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>LOC111099115</td><td>111099115</td><td>1.7885864</td><td>4.161092e-06</td><td>0.0002004183</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>LOC111099216</td><td>111099216</td><td>1.4828614</td><td>7.544561e-04</td><td>0.0125496496</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>LOC111099417</td><td>111099417</td><td>2.7487429</td><td>6.452048e-05</td><td>0.0019567084</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>LOC111099424</td><td>111099424</td><td>3.4384445</td><td>3.211770e-05</td><td>0.0010745946</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>LOC111099722</td><td>111099722</td><td>0.8993515</td><td>1.093514e-04</td><td>0.0028699825</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>LOC111099918</td><td>111099918</td><td>0.6567512</td><td>3.539745e-03</td><td>0.0407210025</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & gene & entrez\\_id & logFC & pval & padj\\\\\n",
       "  & <chr> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & LOC111099115 & 111099115 & 1.7885864 & 4.161092e-06 & 0.0002004183\\\\\n",
       "\t2 & LOC111099216 & 111099216 & 1.4828614 & 7.544561e-04 & 0.0125496496\\\\\n",
       "\t3 & LOC111099417 & 111099417 & 2.7487429 & 6.452048e-05 & 0.0019567084\\\\\n",
       "\t4 & LOC111099424 & 111099424 & 3.4384445 & 3.211770e-05 & 0.0010745946\\\\\n",
       "\t5 & LOC111099722 & 111099722 & 0.8993515 & 1.093514e-04 & 0.0028699825\\\\\n",
       "\t6 & LOC111099918 & 111099918 & 0.6567512 & 3.539745e-03 & 0.0407210025\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | entrez_id &lt;int&gt; | logFC &lt;dbl&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | LOC111099115 | 111099115 | 1.7885864 | 4.161092e-06 | 0.0002004183 |\n",
       "| 2 | LOC111099216 | 111099216 | 1.4828614 | 7.544561e-04 | 0.0125496496 |\n",
       "| 3 | LOC111099417 | 111099417 | 2.7487429 | 6.452048e-05 | 0.0019567084 |\n",
       "| 4 | LOC111099424 | 111099424 | 3.4384445 | 3.211770e-05 | 0.0010745946 |\n",
       "| 5 | LOC111099722 | 111099722 | 0.8993515 | 1.093514e-04 | 0.0028699825 |\n",
       "| 6 | LOC111099918 | 111099918 | 0.6567512 | 3.539745e-03 | 0.0407210025 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene         entrez_id logFC     pval         padj        \n",
       "1 LOC111099115 111099115 1.7885864 4.161092e-06 0.0002004183\n",
       "2 LOC111099216 111099216 1.4828614 7.544561e-04 0.0125496496\n",
       "3 LOC111099417 111099417 2.7487429 6.452048e-05 0.0019567084\n",
       "4 LOC111099424 111099424 3.4384445 3.211770e-05 0.0010745946\n",
       "5 LOC111099722 111099722 0.8993515 1.093514e-04 0.0028699825\n",
       "6 LOC111099918 111099918 0.6567512 3.539745e-03 0.0407210025"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# selecting matching rows in conversion table\n",
    "Entrez <- merge(convert, genes, by = 'gene', all = TRUE)\n",
    "head(Entrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "1e22d65d-dffb-4747-972d-0a0bb2b29dfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>entrez_id</th><th scope=col>logFC</th><th scope=col>pval</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>LOC111099115</td><td>111099115</td><td>1.7885864</td><td>4.161092e-06</td><td>0.0002004183</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>LOC111099216</td><td>111099216</td><td>1.4828614</td><td>7.544561e-04</td><td>0.0125496496</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>LOC111099417</td><td>111099417</td><td>2.7487429</td><td>6.452048e-05</td><td>0.0019567084</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>LOC111099424</td><td>111099424</td><td>3.4384445</td><td>3.211770e-05</td><td>0.0010745946</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>LOC111099722</td><td>111099722</td><td>0.8993515</td><td>1.093514e-04</td><td>0.0028699825</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>LOC111099918</td><td>111099918</td><td>0.6567512</td><td>3.539745e-03</td><td>0.0407210025</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & gene & entrez\\_id & logFC & pval & padj\\\\\n",
       "  & <chr> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & LOC111099115 & 111099115 & 1.7885864 & 4.161092e-06 & 0.0002004183\\\\\n",
       "\t2 & LOC111099216 & 111099216 & 1.4828614 & 7.544561e-04 & 0.0125496496\\\\\n",
       "\t3 & LOC111099417 & 111099417 & 2.7487429 & 6.452048e-05 & 0.0019567084\\\\\n",
       "\t4 & LOC111099424 & 111099424 & 3.4384445 & 3.211770e-05 & 0.0010745946\\\\\n",
       "\t5 & LOC111099722 & 111099722 & 0.8993515 & 1.093514e-04 & 0.0028699825\\\\\n",
       "\t6 & LOC111099918 & 111099918 & 0.6567512 & 3.539745e-03 & 0.0407210025\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | entrez_id &lt;int&gt; | logFC &lt;dbl&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | LOC111099115 | 111099115 | 1.7885864 | 4.161092e-06 | 0.0002004183 |\n",
       "| 2 | LOC111099216 | 111099216 | 1.4828614 | 7.544561e-04 | 0.0125496496 |\n",
       "| 3 | LOC111099417 | 111099417 | 2.7487429 | 6.452048e-05 | 0.0019567084 |\n",
       "| 4 | LOC111099424 | 111099424 | 3.4384445 | 3.211770e-05 | 0.0010745946 |\n",
       "| 5 | LOC111099722 | 111099722 | 0.8993515 | 1.093514e-04 | 0.0028699825 |\n",
       "| 6 | LOC111099918 | 111099918 | 0.6567512 | 3.539745e-03 | 0.0407210025 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene         entrez_id logFC     pval         padj        \n",
       "1 LOC111099115 111099115 1.7885864 4.161092e-06 0.0002004183\n",
       "2 LOC111099216 111099216 1.4828614 7.544561e-04 0.0125496496\n",
       "3 LOC111099417 111099417 2.7487429 6.452048e-05 0.0019567084\n",
       "4 LOC111099424 111099424 3.4384445 3.211770e-05 0.0010745946\n",
       "5 LOC111099722 111099722 0.8993515 1.093514e-04 0.0028699825\n",
       "6 LOC111099918 111099918 0.6567512 3.539745e-03 0.0407210025"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>405</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 405\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 405\n",
       "2. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 405   5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# grabbing only unique IDs\n",
    "gene_entrez <- Entrez[Entrez$gene %in% unique(Entrez$gene),]\n",
    "head(gene_entrez)\n",
    "dim(gene_entrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "a6d5dc7f-97f2-4377-a5c4-513e8796121e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>111099115</dt><dd>1.78858640472996</dd><dt>111099216</dt><dd>1.48286142511512</dd><dt>111099417</dt><dd>2.74874287789152</dd><dt>111099424</dt><dd>3.43844453272146</dd><dt>111099722</dt><dd>0.899351501638737</dd><dt>111099918</dt><dd>0.656751219091182</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[111099115] 1.78858640472996\n",
       "\\item[111099216] 1.48286142511512\n",
       "\\item[111099417] 2.74874287789152\n",
       "\\item[111099424] 3.43844453272146\n",
       "\\item[111099722] 0.899351501638737\n",
       "\\item[111099918] 0.656751219091182\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "111099115\n",
       ":   1.78858640472996111099216\n",
       ":   1.48286142511512111099417\n",
       ":   2.74874287789152111099424\n",
       ":   3.43844453272146111099722\n",
       ":   0.899351501638737111099918\n",
       ":   0.656751219091182\n",
       "\n"
      ],
      "text/plain": [
       "111099115 111099216 111099417 111099424 111099722 111099918 \n",
       "1.7885864 1.4828614 2.7487429 3.4384445 0.8993515 0.6567512 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gseaDat <- gene_entrez\n",
    "ranks <- gseaDat$logFC\n",
    "names(ranks) <- gseaDat$entrez_id\n",
    "head(ranks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d2bb0a83-c3e8-44b6-9339-70c533fdf1f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OU\nlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWm\npqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4\nuLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnK\nysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u\n7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////i\nsF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dB5wb5bmo8XdUt0i2qaGZxRjc\ne2dNgADGh3ANOCHBhF5CJwW4l2IMIdScgAk1GBKaISSGGwIcQlmCc0Ix4KWaOAZj7IRifCjB\nBWzjojMz0hbtStrd0futZqTnD7+VZqSRvvl2Hu9Kq9VKCkDRpNQDAMoBIQEKCAlQQEiAAkIC\nFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkIC\nFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkIC\nFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkIC\nFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkIC\nFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkIC\nFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCgoZUifN75XwnsHFJUypJNl\nmxLeO6ColCEdl+xZwnsHFBESoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCgg\nJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCgg\nJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCgg\nJEABIQEKCAlQUNqQamb9q4T3D6gpbUhiTS/h/QNqShpSQpIXlvD+ATWEBCggJEABIQEKCAlQ\nQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAUlDqm617A1JRwBoKTE\nIcXj8mEJRwAoKXVISUJCOSAkQAEhAQoICVBASIACQgIUEBKggJAABYQEKCgypM8fmHnbW143\nJiSUDa8hbb74Afvj7QmxHbba222kQ7rwWY9DAPzDa0jz5aZU6jHLmvSTY7aTw7zdRjqk0FSP\nQwD8w2tIs+S5VGpi+En77Kq95U1Pt5EOKTl5qccxAL7hNaQrZGFqczT9pehl56tTttWft8j7\nexLpkGpEnvc4CMAvvIZ0s8xNfW2d457/RK5sc+m7lrSwNuS5jXRI1dXhBo+DAPzCa0ivypmp\nVL9vuecfl7vaXrygsdkdsj7PbWRCqiEkBJ7np7/3DN2Zuk2utL/aLOgbX17gis93GJK1/TFe\nRwH4g+eQlmwvIy4YKtvuOTgs1xS6YidCivf3OgrAH7z/QPbDQzIPhLa/u+D1OhFSkpAQcMW8\nsmHJbeee8pNfPrOx8LUICRXA/GvtCAkVgJAABYQEKCAkQAEhAQoICVBASIACQgIUEBKggJAA\nBYQEKCAkQAEhAQoICVBASIACQgIUEBKggJAABYQEKPBHSH1ey/cmkkAg+CKkhCW3Gx8HYJAv\nQqqVxA3GxwEY5JOQkoSEQCMkQAEhAQp8ElLN/v9pfCCAOT4JKRaRTcZHAhjjl5CShIQgIyRA\nASEBCggJUEBIgAJCAhQQEqCAkAAFhAQoICRAASEBCggJUEBIgAJCAhQQEqCAkAAFhAQoICRA\nASEBCggJUEBIgAJCAhQQEqCAkAAF/gnph08bHwpgin9CipxmfCiAKf4JKXnwAw3vGx8NYIR/\nQqoSkUnGRwMY4aOQ7C9K+xgfDWCEr0JK7HjKMuPjAQzwVUg1Ep1jfDyAAT4LKUlICCSfhZSY\n8bbxAQH6fBZSVHY0PiBAn99CSm5rfECAPkICFBASoMB3IW3VuNb4kABtfgspIbLj3vk2APzK\nfyHF4nLMo8ZHBajyX0j2qsg046MCVPkypCQhIWAICVBASIACQgIUEBKggJAABb4MKTFi5lfG\nxwUo8mVI1SF5zvi4AEX+DCkpO5xifGCAHr+GFB9ufGCAHr+GlCQkBAkhAQoICVDg15AStaP/\nZnxogBa/hlQr8VnGhwZo8W9ISUJCcBASoICQAAWEBCggJECBf0OqPfmvxscGKPFvSFGr2vjY\nACU+DikZMz42QAkhAQoICVBASIACP4ckcoXx0QEqfB1S9dhL+MMUCARfhxSPyNixR75vfIhA\nsfwdUlIiVuQR40MEiuX3kGrCsYknNaw2PkqgKP4PKRwVudb4KIGiBCCkhCSvNj5KoCiEBCgg\nJEBBIEJKnDTf+DCBYgQipJhYa42PEyhCMEJKygcbjA8U8C4oIcmPjA8U8C4oIdUeZ3yggHdB\nCSkx9sp5K4yPFfAoKCFVhUS+aXysgEeBCcn+ojTO+FgBjwIUUpKQ4FuEBCgIUEi1ochTxkcL\neBKgkGqk+m7jowU8CVRISUKCTxUd0sr3Nxa+AiGhAhQd0iXSwZuTEBIqgNeQNm3ImCHLNmzY\nVOCahIQK4DWkS6S1Swpck5BQAbyHlOzpikuPnj2vKnBNzZCu+czjcAGzvIb0UHzsQvdMzsdI\nq6af1+wovZBisp3H4QJmeX6y4c3hVb90HhrlDGnFtO8121cvpGht0utwAaO8P2u3/rxQ/Tvd\n+6xdNFn79CrPAwbMKebp72f71NywuXtDEplRxIABU4r6OdKqE+Vbx3VvSMnzixkwYEiRP5B9\neFshJKDoVzb8z6XTVxa+hm5I1dtNWMBbc8F3AvWiVWdVXORI42MGuihwIdkXHWJ8zEAXBTCk\nJCHBdwgJUEBIgAJCAhQQEqCAkAAFhAQoCGJI+7yx2fiogS4JYEg1IlNuMj5soCsCGFJ13Irs\nbnzYQFcEMaQqK0lI8BdCAhQEM6RE7fAx+976+hrjgwc6J5gh1Uo4JpaElxkfPdApQQ3JWVUj\nh91nfPhAZwQ5pKRE9jI+fKAzgh1SkpDgD4QEKAh2SIkeY18zvgNAx4IdUo3Ef298B4COBT2k\nJCHBDwgJUEBIgIKghxSTYcb3AOhQ0EOKVm9tfA+ADgU+pCQhwQcICVAQ/JAs2bn3vrOM7wZQ\nSOBDSkg0LrGxN7zwsfE9AfIqg5CcVZbIjue/Z3xfgDzKJCTnbGjricZ3BsitbEKyN6iKLNlg\nfHeAXMoppKTIOauN7w+QQ3mFFJfIh8Z3CGivvEKqqpG3je8Q0F6ZhZSUO/mShBIou5Bkz1c2\nGt8noI2yC6lGZLfjjO8UkK38QrIi8fjI63mohG5VjiElJCLWfxzwo5XG9w3IKM+Q7EsjETn0\nNuM7B6SVbUj22Whi3Hzjuwc4yjkk+9KxRzd8anwPgXIPKSrWBcb3ECj3kJKS/H/G9xCogJCO\nWWh8F4GyDykuUnfsY+98bnw/UdHKP6SkREVkqPH9REWrhJCq46Havsb3ExWtQkJKEhKMqpCQ\nEpYcfrvxXUXlqpCQ7EtD8vY64zuLSlUxIdmrZNzFy/ldJRhRSSHFIyJ9Tl1qfI9RgSopJHtV\nlYSHHPnKWuM7jUpTYSHZq2IiZywyvteoMJUXkr2BWLzQAboqMqSEyKS7840K8KAiQ0pKzJKZ\ni43vOipHhYZkbyDDje86KkflhpTs85evjO88KkXlhlQrUtNj6CE/eXzhF8bnAGWvkkOKVTnv\nNeTY+vAXjc8DylpFh5RZFbaDGrLfVc/wJ2HgGSHViLNBVCyxvnfWHxZ8YnxCUI4IKR2SMw5x\nvs+L7/+o8SlB+SGk5pDc27DX/nC58TlB2SGk7JCc17VaU+4wPisoM4TUNqR4rUR22PMHJ97S\n8K7xuUHZIKR2IdlnQ5bzcCm0373GZwdlgpByhRR1NrDXnmh8dlAmCClvSPZF2+wxf5XxCUI5\nIKQCIVXFRSYZnyCUA0IqFJJ9aa+9Zz3W+KXxWULAEVIHIUlYRA6f1fCh8YlCkBFSRyHZG8Qs\nkViPE+8zPlcILELqREjunUpkoPG5QmARUidDqpXkrq/wPg/Ig5A6HVKtSN1k49OFYCKkLoQU\njcucZcYnDEFESF0IyV5lnWV8whBEhNS1kJJnGJ8wBBEhdTGkk/iFdORASF0LqVqswf/xn58a\nnzUEDCF1LaQqe4VYfX9gfNoQLITUxZDcDeLhHhNO5W+WoQUheQkpIZGo8Bct0IKQvIVknyUk\ntCAkQoICQiIkKCAkzyH1+Oapd/+x8WvjE4ggICTPIUUjYonU7t6wjD8PA0LyHFJ6gyoRu6cd\n6y+/8oHPjM8lfIuQigzJ3qBG4pZliTV8xLcnT730GeMzCh8ipOJDss9Gwu5ftAhHJNT/QV6N\nV4EISS+kzKXWtN/+2/iswmcIST2kaMSSo58wPq/wFULSDykhUUuSAw75Pe/CXzkIyURI9qpY\nRGSrQ0/4+RLeMKUiEJKhkNzdCtk1jfgpP2aqAIRkMiR7VTwiC4zPMUqOkAyHZK/absvdG41P\nM0qLkMyHFJXI0Cn3v/SR8alG6RCS+ZDsVTGx7P9qEjvufABfnMoSIXVLSPYq+wpVEpPI1qMb\nlvBnYsoOIXVXSE0bxO0vTaG9Zhqfd3QrQurukNxXucZ4E/EyQ0glCCkpid7Tfv3npbwPUfkg\npJKEVOP83oVI8uY3jc8/ugUhlSakkL1BtFpC9TNueeS9zcY/CTCNkEoXknOp86y47HHKknXG\nPw8wipBKG5J7czGRY583/omASYRU+pCcVZbstNcFvDd/cBGSP0JKSjRsTX/Z+GcDhhCSX0Ky\nd8vqb/yzAUO8h7Rkxgm/TH8v8vHk2QWuR0idDim5zfSb7m7gDR+CyHNIz1SLyBaPOGeXyiUF\nrkhInQ6pVkKWyJlePyUoIa8hfd1bDrv97OrQXSlCUgzJXlU7cNJPH/D4WUHJeA3pSTna/rho\nt9B9OUNa83mzJwipSyFVRSWyxcRrn/X4iUFpeA3pWpnrnKwYFH4gR0jvhqSVfD9sJKS8uxW2\nev7qtt88/oeHXv7A4ycI3ctrSFfJ393Tj3aLPprjK9Kbjc3u4CtS10OyN7DcFz1IYvexe1+9\n0ONnCd3Ga0j3ypPpM//sHb+Nx0gGQmraIByzm6q+6M5XPuWNvXzMa0hvyXmZc29/I0xIBkNy\nVlWL871yjwkHX8TrW33K89Pfu+/U9Ce23tySkAyH5O5WtfMe/YP2OvLupxe/y18L9BvPIT11\n3YdNZxvPayhwRUJSCsndwPnrZvZDp5B1vddPHIzgJULBCqnKWRWplfguBz7/6lLjnzx0FiEF\nMCRng5jzg4UpP3+Nv2HrD4QU0JDsS2NWRGTAbntOvfgj3v2h1AgpuCG5uxWTiJ1TKHzxNXMW\n8W79pUNIAQ/JXVUlYefHt+HeEw6ecX0jP28qAUIqh5DSG9Q4z+mFRQ682fgnFW0RUvmEZK9K\nOBtErAZ+ztTdCKn8QkqKVCXq9zzoyGu/MP7ZRQYhlWNIVdUSFStiyc59Dpx82tJ/E5RxhFSW\nIWVGGY07LysSS4ZNvubqm+f/7Q1eqmcKIZVzSOlRVjnvnOe+RbJIzU7fPW7W28Y/6ZWHkMo/\npFa75b7w1apKDB556lLjn/mKQkgVFVL6Npw/eBY+odArjdFVhFSBIbkbWOFe5z7WyHt/KSGk\nSg0pIRH7MVPdd+a8+N5n/CnOohFS5YZkr6qKuE9BRPqczZv4F4eQKjokd5TRWueL0zvGj4Sy\nRkiElF71naNn/PEt40dD2SIkQkqvCjt/bd0atOteh13DQ6auIyRCarVBNO78ehOPl7qOkAip\n1QbpO91+1Kk33vpkI2/y2gWEREjtQoo5fxXD/u+Fvy81fniUC0IipHYhpTdIOE+Mj5p8/pVX\n3jfnb7zatQOEREh5QrLPhkPRpvcgj/c6eebL/1hh/HAJKkIipPwhtexWlfv7GBL5zPjxElCE\nREidCcndIFYrPQ74Le9VlAshEVKnQ7JXxSxrxl0NHxJTW4RESF0Jqbra/Q3B6P43LTd+5AQK\nIRFSF0OyN7CvK6Ebbvj1vQ82Ll9j/AgKBEIipK6HZK+qdZ7Lc744Wfa65Pb7v2T8QPI3QiIk\nTyG1PP8Qq5K4RAfutu/MBcYPJv8iJEIqJqTmDWLOr2Ic8fNnNhg/oPyJkAhJJSR3g7DIWU8b\nP6J8iZAISS8ke7esXqdecumcNyvuNzEIiZBUQ0pkXvC64+CTK+p3bgmJkJRDci6tcd7wa5ud\nT37W+OHlF4RESAZCci+tksiIvY877fJ7rvv9q58YP85KjJAIyVRI9qpo+i+giVQ9Mb+8XwpB\nSIRkMCRnlM4G7o9ve06dt7hsvzIREiF1Q0hJ56UQzh9i7238eCsRQiKk7gkpfQ+RHcYPn/KD\na8vvBXqEREjdF5K9QSQm4ZBIvHb4wO/dU0Zvr0JIhNStIaU3qJJI3HknvVvK5mdNhERIJQgp\nvSoh1pAbn33f+CHYHQiJkEoWkrNKpHbkqEPfyXeMBAYhEVIpQ7IitRKOifQaEvAvTIRESKUN\nyd0gHpdRow9vCPB7FBESIfkgJPtsVEKy9ZJ1xo9HQwiJkPwRUq3zd5pk4sn3LzZ+SJpASITk\nm5DsDSJiffsL48ekAYRESH4KybmW7HCU8aNSHSERkr9CskcZ3/a2j40fl8oIiZB8FlLMHqU1\n4qBb/2H80NRESITku5DsS6NixWsG7nbAC2uNH6A6CImQ/BiSvapaohKWsGXVHHnsxY8uMX6g\nFoeQCMmnIbkbVDv/Oy8XD1lnz2708fN5hERIfg6pZbcilsiwI2YvMn7AekNIhBSMkJxV9iOn\n0U997MtXPxASIQUnJGeVSO/7ffhGKoRESMEKKVItloQf/dD4gds1hERIAQsp4b4pUf8z/m38\n0O0KQiKk4IXkDE2eN37odgUhEVIgQ0rKE7560oGQCCmgIUm8NtGrX99vrTZ+CHcGIRFSQEOq\nlrD7ZkQ7X2D8GO4EQiKkgIbUtEF86O0++KJESIQU8JBqLOt7dxs/jDtCSIQU8JCqI6FIJJEY\ntNH4oVwIIRFS8ENKSKRKbnrb+LFcACERUjmEZK+y+hzxufGjOS9CIqQyCakqJtttN/KWEr08\nnJAIqVxCSjp/uFa2MX5E50RIhFQ+ITl3Gkpss/vDT7xh/Lhug5AIqaxCcp93sESSgy7o1t9O\nJyRCKrOQ0rsVl/DE6cYP7haEREhlGVKt805Em4wf3c0IiZDKNaQkIRESIRFSG4RESISkgJAI\nqVQhnTX9rg3GD/A0QiKk8g0pZMlOE6avMH6MpwiJkMo5pCrnDwFGHjR+jKcIiZDKPKQaqdp1\nn1nLjB/mhERIZR5SJC6hy4wf5oRESOUeUkKSlxo/zAmJkAhJASEREiEpICRCIiQFhERIFRDS\nBV+ZPswJiZDKP6S4VK8xfJgTEiFVQEg18qnhw5yQCKkCQkoSEiEREiE5CImQCEkBIRESISkg\nJEIiJAWEREiEpICQCImQFBASIRGSAkIiJEJSQEiEREgKCImQCEkBIRESISkgJEIiJAWEREiE\npICQCImQFBASIRGSAkIiJEJSQEiEREgKCImQCEkBIRESISkgJEIiJAWEREiEpICQCImQFBAS\nIRGSAkIiJEJSQEiEREiFzDt334F1A/c998XCVyMkQiKk/FYeLE0OXlnoioRESISU3/4y8oZ5\nyz5eNu/64TKp0BUJiZAIKa/n5ZhNmbMbj5J5Ba9JSIRESHnMlPeazy+RX7W5dNVF5zU7Kn9I\nMQnHJBKRqHsas09DsbZnLavdpZH2q2Lp23A3CLW7NJx3A/s0Iu03CLXfINrlURbeoP2dFt6t\nHKPMsVtZo+zO3co/+fl3K8/kd7Rb7UbZqd066QffP/O7R5966MnHfuesw484c+rxJx163n97\nPPhz8RrS1a0S/0R+0ebSFVP2bzZW1uW5jfv36z929IA9hg6pHzBqXL/xIwbWDx62R/8xY/rv\nMWxw/cAR4/uNGzWwfsjQPQaMHtt/wvDBEweNmNBv7KgB9U0bjBzUaoOJA0eO7zd+lHMb9f1H\nj+k3YfigeneD0QPqh6Q3mDBi0MTBw51V/Z07HThyXL9xIwdOHOJsYN+De6cT7Du172FovX2n\nzgb1g4fv0W+MvcEwe4NR9j2MbBnlcHuDkePdDdKj7DfeuVN7g07slr3KuYdcu9W8Qb1zp849\nOKuGZu3WuMxujXZ3y77TCe4ohw5utVv2bWR2q3mU9j3k3q1co2zerQktu1XfereGt5t8d7fG\nZO/W0FbzYK9q2q1xLbvlzkN6t0bl3a02kz8se/Jb7VZm8gdmTX5VOCJhCUWc/8OW5Z5GjvN4\n8OfiNaSH5Prm89fJwwWumf9bO6C7bJvj4ULSDyGt3Tly3mL33DvnhuvWFrgmIaH0fBtSav42\nIr0GjRzUS2SbxkJXJCSUnn9DSn16YX9LRKz+FxZ+QoSQUHo+Dsm2ZvGbizv8a9GEhNLzd0id\nQkgoPUICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQ\nQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQ\nQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEDBtnFCAoo2\nUAgJKNrXdxISULw5hAQUj5AABYQEKCAkQAEhAQoICVBASIACQgIUEBKggJAABYQEKCAkQAEh\nAQoICVDw/y1CAor25YNDCAko3j6EBBSPkAAFhAQoICRAASEBCvaJERJQtLN3IiSgeAsICSge\nIQEKCAlQsEAihAQUa1lYCAko2poTCQko3mmEBBSPkAAFp0UICShaww8JCSjeJkICikdIgAJC\nAhQQEqCAkAAFhAQo2BwRQgKK9vrpNVJNSECR7hFLCAko1hfLbiEkoHgNhAQUj5AABYQEKCAk\nQAEhAQoICVBASIACQgIUEBKggJAABYQEKCAkQAEhAQoICVBASIACQgIUNIRChAQUa74IIQFF\nWzWFkIDiTSUkoHiEBCjwV0hf/WtD+sy6ee8XuBohwW/8FNK6U2LS6+KvnbNL5ZICVyQk+I2f\nQjpeZMeY1H+eIiQEjY9CWmrttCD16Qky5gtCQtD4KKQ75HbnZKbssZqQEDA+CukKed09/ZXs\n9SUhIVh8FNJN0pg+c4VMWkRICBQfhfRn+U3m3Pkytn1Iaz5v9gQhwWd8FNLqqm83nT1d2oX0\nblha+drz6AATfBRSatqWSzPnNh/b/ivSG40t/ur1LgAz/BRSK5vXblC4FaC7+DQkIFj8F9LK\n/e7WGQfQffwX0icFn/kGfImQAAWEBCiYGiMkoGg3ftNvIX05/S864wC60Qd+CwkIIkICFBAS\noICQAAWEBCggJEABIQEKCAlQQEiAAkICFHwSEkICirb4REICinchIQHFIyRAASEBCggJUDDD\nIiSgaJ81bCEnK94eIaFCvdf4ueKtERKggJAABYQEKCAkQAEhAQoICVBASIACQgIUEBKggJAA\nBYQEKCAkQAEhAQoICVBASIACQgIUEBKggJAABYQEKCAkQAEhAQoICVBASIACQgIUEBKggJAA\nBYQEKFAKaePCuc8XtWB7Ze6rmXOb3n/xnbVNq7947aVPWq70rxcWbS44ks+em/thFxc2Ln1h\n4cbWNzF3rvuu0Ovnpv09s37VgsYVzVfKWsil6ClptfB2ZiRfphezpiRrsnLyMCXZC2tfe7n1\n22Q3zc+izKjcz1vbycrFyyzk39cNC174ZyrnQsdTok8lpPsn1Ih8o4gFx3Mhmeie2Xzd9iJS\nfeqn9vkvLhpsn5c956ev8/o4e6HukbwDWT+1zr7Cr7u08MRhtfbCFjPWN9/KZJEG53S5pB3u\nrl14iGWfH/bH9gtGpiRr4ZTMSP7RbkpaT5balGQtpDbNSIhEjvqi3fycmBmVTGo3WUpTUmBf\nN13Ry14YMS/VbqHDKTFCJaRzeu1zRvNEeFmwrd5110xIF0jdNQ/NGiuj7S89r8muZ9xw1V5S\n3ehcsmTL8LkP/XKL8BP5BrI61H/ahKYjoJMLh8S+fdlNP+4h32+6kZtCdU0hDbnT8Vd3YUs5\n+t4Hzq2Sh9sumJmSrIVT5CJ3JM7BnD0lrSdLbUqyFlI/luF3zjlI9trUdn6ed8d05yC5ve1k\naU1JgX09TXpf9+CF1TXudzFZCx1OiREqIa22/2+eCC8LthN7PpoJqYf1nv1x7c7yWir14ZPu\nqtPlYOfk+3Kd/fF52T3fFG1eY38imo6ATi486n7nsLiHZL6xfLvmnEOaQprcctMzxf2zVDfL\nIW0XzExJ1sIpMq/5lrOnpPVkqU1J1sJbod4r7VWT5d6285O2Ipa0r549WVpTkn9f/2Ft4Xxn\n/bh8M9VmoeMpMULryYasKDwsPCJ3vZYOaWOk2v23by95o/nSt6Sv/XFVvMe69CVZj63aaD4C\nurBgO0xmu6cbxg1Ymyukn8lVzskzMrXtQj7FTkmekDLSU5JrstrzMiXNCxc07eskd7HV/KRd\nLac5Jx2HlPI2C6mc+3qLHO9etIu822ahc1Oizich/c83pqQyIaW+JZfbD/3/FB2/qfma86Te\n/vjXzOdyhlxbYCQeQ5osT7mnPwu/lGoOaYcJfUafsdBd/9/S1/4sfTZZ/qvtQj6qIX2zX78D\nb/+61a2npyTXZLVXVEj7pP/ZWh/u4S62mh/X5r7pf/mzJisfjyHl2NdfyLnuRaOdf/+yFjo3\nJep8EtKhWy5vDmnJENlyxI7J01oe3aYOldvsj3fKD92l38iZBUbiLaRXwru4zzbMj1yYagmp\n57ARCYnd417j8ki435CqMY+0X8hDM6RQn/E7WzKh3ZTkmqz2igqpTtJP3+0kzqP31vPjekrG\nu6fZk5WHx5By7Osc2cdZtSohl7ZZ6NyUqPNHSHfI/anmkFJP9dtqXO/QAUuar3iV7O/863Kj\n/NRd/L0U+iu6nkJavkv4GXjuKTcAAAVTSURBVOf0qwFD1zcfKF/+xf6Xbd1lElnkLH320+iA\nEdXbpb8YZi3koRjSSx/ZH14dKkc3X5aZkhyTlUNRIW0lK93TAbKszfy4viN3uKfZk5WHt5By\n7evqreUG+06PEOeLUdZC56ZEnS9CWpr8bqolpJ/JZRtSqYdj2/07c70bZZT7c4tb5Sx3eXbB\nv6Lr5aj5eHAofTicGXWecsh6DJA6Xs6zP/6zrr99iHx6kPuIIWshH8WQ0paEoyszZ5umpP1k\n5VJUSNvJZ+5pX/kox/wsj/T8svVdpScrH087nntfn07K9mN6VE+RGW0XOjUl6nwR0tTonLlz\n594uQ+a+mEq9ExrjrvyR+xRdyvkOeGx6Ih+Uae7pdXJBgZF4OGre7xe6yz3zunWo80PFiXLN\n3I+bL31Aptgfj04/HvrU6tV2IR/1kFKD5ZX0meYpaTdZORUV0hBJf43pGVqbY34ub/Ntdnqy\n8vGy4/n29YNLD5589uLT09/1tVro3JSo80VIE5t+suc8O/M7Ocpdeauc4Z5eJBMz/w4vksHu\n6QlyX4GRdP2oebcucn/6XEPzQDLP4Tlmy6H2x37OM0Ip5x/oT9os5KMfUl953T1tmZK2k5Vb\nUSFNkweck3/J7jnmZ9Mu8mbWXaUnKx8PO97Bvu4uC9ssdG5K1PkipCdnO66QfrP/ZH9NlpHu\nytPl5/bHzT+Sfdc0bdDXnbV120YKHL5dP2re2j72UObscncgs0fL+bPfa77mwfIz++NI+ZOz\nsMKKf91mIR/1kBqt6q9S2VOSPVn5FBXSPekXK8yUH+eYnz/LHtl3lZ6sfLq84x3t6+9kz1Sb\nhc5NiTpfhJSWeYy0sqdMtx/Q/iEaesv+F+8EOXBt8zVmyYRPUuuPL/gQqcsHSuNW1Y+3uYnM\nY4Cr/2B/l/3PEyXxQcr5zruv/W/vx5PcwyprIR+9kBad/9aG1Po/9nafbMmakqzJyquokNb1\nse5JpV7uFW/5p6XlMdIhclfmXNZk5dPVHS+wr0+/uDH15azquPs1uvVC56ZEnUpIz9XV7Syh\nurq69z0uuJqebHi4RnoM21YiN6acJ1dlhzpHP+eSzSdI9YgtZcLqvCM5rK6uh2xZV3dO5xfq\npca9h7pbmm8lc6AcJFKbFNn6L87CV5Mk1GdgTIataLtgZkpaL8wXCW0dEpm2vt2UtJ4stSnJ\nXnhlK9llsBX9XctNNof0QbjXV5l1WZOlNSUF9vUcCW8bla3SA8la6HhKTFAJqXHvjOUeF1yL\n9848bn3/qu/ud/B093vfl5qulf5RbOqx4w84/Lcb8o/kzMzVr+r8wmlNd9HywGv63u5ru965\n5aT/M+moX2d+HrH54VMmTzpu9vr2C0ampPXCuv86//D9p/zfF1I5pqTVZKlNSfZCasVlUw48\nu/U9ZOYnlZqz9y+a1mVPltKUFNjXJb+YNun712delJ610PGUmMDvIwEKCAlQQEiAAkICFBAS\noICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBAS\noICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBAS\noICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBAS\noICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCggJEABIQEKCAlQQEiAAkICFBAS\noICQAAWEBCggJEABIQEKCAlQQEiAAkICFBASoICQAAWEBCj4X4Z9Jh8Kc/ICAAAAAElFTkSu\nQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "barplot(sort(ranks, decreasing = T))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df90f8c6-e7eb-452d-ab5f-0ec5a438f597",
   "metadata": {},
   "source": [
    "#### load pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f01794ba-d513-45aa-a560-471a93814df9",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(KEGGREST)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c07c6e0f-b24b-4126-a7ef-ac09e0366935",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$cvn01100</dt>\n",
       "\t\t<dd>'Metabolic pathways - Crassostrea virginica (eastern oyster)'</dd>\n",
       "\t<dt>$cvn01200</dt>\n",
       "\t\t<dd>'Carbon metabolism - Crassostrea virginica (eastern oyster)'</dd>\n",
       "\t<dt>$cvn01210</dt>\n",
       "\t\t<dd>'2-Oxocarboxylic acid metabolism - Crassostrea virginica (eastern oyster)'</dd>\n",
       "\t<dt>$cvn01212</dt>\n",
       "\t\t<dd>'Fatty acid metabolism - Crassostrea virginica (eastern oyster)'</dd>\n",
       "\t<dt>$cvn01230</dt>\n",
       "\t\t<dd>'Biosynthesis of amino acids - Crassostrea virginica (eastern oyster)'</dd>\n",
       "\t<dt>$cvn01232</dt>\n",
       "\t\t<dd>'Nucleotide metabolism - Crassostrea virginica (eastern oyster)'</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$cvn01100] 'Metabolic pathways - Crassostrea virginica (eastern oyster)'\n",
       "\\item[\\$cvn01200] 'Carbon metabolism - Crassostrea virginica (eastern oyster)'\n",
       "\\item[\\$cvn01210] '2-Oxocarboxylic acid metabolism - Crassostrea virginica (eastern oyster)'\n",
       "\\item[\\$cvn01212] 'Fatty acid metabolism - Crassostrea virginica (eastern oyster)'\n",
       "\\item[\\$cvn01230] 'Biosynthesis of amino acids - Crassostrea virginica (eastern oyster)'\n",
       "\\item[\\$cvn01232] 'Nucleotide metabolism - Crassostrea virginica (eastern oyster)'\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$cvn01100\n",
       ":   'Metabolic pathways - Crassostrea virginica (eastern oyster)'\n",
       "$cvn01200\n",
       ":   'Carbon metabolism - Crassostrea virginica (eastern oyster)'\n",
       "$cvn01210\n",
       ":   '2-Oxocarboxylic acid metabolism - Crassostrea virginica (eastern oyster)'\n",
       "$cvn01212\n",
       ":   'Fatty acid metabolism - Crassostrea virginica (eastern oyster)'\n",
       "$cvn01230\n",
       ":   'Biosynthesis of amino acids - Crassostrea virginica (eastern oyster)'\n",
       "$cvn01232\n",
       ":   'Nucleotide metabolism - Crassostrea virginica (eastern oyster)'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$cvn01100\n",
       "[1] \"Metabolic pathways - Crassostrea virginica (eastern oyster)\"\n",
       "\n",
       "$cvn01200\n",
       "[1] \"Carbon metabolism - Crassostrea virginica (eastern oyster)\"\n",
       "\n",
       "$cvn01210\n",
       "[1] \"2-Oxocarboxylic acid metabolism - Crassostrea virginica (eastern oyster)\"\n",
       "\n",
       "$cvn01212\n",
       "[1] \"Fatty acid metabolism - Crassostrea virginica (eastern oyster)\"\n",
       "\n",
       "$cvn01230\n",
       "[1] \"Biosynthesis of amino acids - Crassostrea virginica (eastern oyster)\"\n",
       "\n",
       "$cvn01232\n",
       "[1] \"Nucleotide metabolism - Crassostrea virginica (eastern oyster)\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pathways.list <- as.list(keggList(\"pathway\", \"cvn\"))\n",
    "head(pathways.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "97f84b3d-789d-46d6-a86b-6feed34c1440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "136"
      ],
      "text/latex": [
       "136"
      ],
      "text/markdown": [
       "136"
      ],
      "text/plain": [
       "[1] 136"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "136"
      ],
      "text/latex": [
       "136"
      ],
      "text/markdown": [
       "136"
      ],
      "text/plain": [
       "[1] 136"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(pathways.list))\n",
    "length(pathways.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "cde724b7-a746-4bff-b948-7c84bee1bf7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>111099115</dt><dd>1.78858640472996</dd><dt>111099216</dt><dd>1.48286142511512</dd><dt>111099417</dt><dd>2.74874287789152</dd><dt>111099424</dt><dd>3.43844453272146</dd><dt>111099722</dt><dd>0.899351501638737</dd><dt>111099918</dt><dd>0.656751219091182</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[111099115] 1.78858640472996\n",
       "\\item[111099216] 1.48286142511512\n",
       "\\item[111099417] 2.74874287789152\n",
       "\\item[111099424] 3.43844453272146\n",
       "\\item[111099722] 0.899351501638737\n",
       "\\item[111099918] 0.656751219091182\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "111099115\n",
       ":   1.78858640472996111099216\n",
       ":   1.48286142511512111099417\n",
       ":   2.74874287789152111099424\n",
       ":   3.43844453272146111099722\n",
       ":   0.899351501638737111099918\n",
       ":   0.656751219091182\n",
       "\n"
      ],
      "text/plain": [
       "111099115 111099216 111099417 111099424 111099722 111099918 \n",
       "1.7885864 1.4828614 2.7487429 3.4384445 0.8993515 0.6567512 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Assuming geneList is your vector or data frame with gene names\n",
    "\n",
    "head(ranks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "a25667db-59aa-4cc5-b18e-726e98315052",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are duplicate gene names, fgsea may produce unexpected results.”\n"
     ]
    }
   ],
   "source": [
    "fgseaRes <- fgsea(pathways.list, \n",
    "                  ranks, \n",
    "                  minSize=15, \n",
    "                  maxSize = 500) \n",
    "                  #nperm=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "fbe029de-d15a-4a34-9db0-125e6cdc73b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th><th scope=col>leadingEdge</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 0 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " pathway & pval & padj & log2err & ES & NES & size & leadingEdge\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & <list>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 0 × 8\n",
       "\n",
       "| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; | leadingEdge &lt;list&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     pathway pval padj log2err ES NES size leadingEdge"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fgseaRes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5956864-c939-4ca5-9d5b-7587fb19cc10",
   "metadata": {},
   "source": [
    "#### GO enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "6c01ba95-e549-4714-b916-46ec3bc7dd20",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'fgsea'”\n",
      "Old packages: 'BH', 'biomaRt', 'Biostrings', 'brew', 'brio', 'bslib',\n",
      "  'commonmark', 'cowplot', 'curl', 'data.table', 'DBI', 'desc', 'digest',\n",
      "  'GenomicAlignments', 'ggfun', 'ggrepel', 'glue', 'GOSemSim', 'graphlayouts',\n",
      "  'haven', 'htmlwidgets', 'httpuv', 'igraph', 'MASS', 'Matrix', 'mgcv', 'nlme',\n",
      "  'patchwork', 'pbdZMQ', 'pkgbuild', 'pkgload', 'ps', 'ragg', 'Rcpp',\n",
      "  'RcppArmadillo', 'RCurl', 'readr', 'reprex', 'Rhtslib', 'rlang', 'scales',\n",
      "  'shadowtext', 'shiny', 'SparseArray', 'testthat', 'tidygraph', 'tidyr',\n",
      "  'timechange', 'uuid', 'viridis', 'vroom', 'withr', 'XML', 'xml2',\n",
      "  'yulab.utils', 'zip'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#BiocManager::install('fgsea')\n",
    "library(fgsea)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "60b04c06-23a8-42d2-b913-151a4e8dab3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "GSEA = function(gene_list, GO_file, pval) {\n",
    "  set.seed(54321)\n",
    "  library(dplyr)\n",
    "  library(fgsea)\n",
    "\n",
    "  if ( any( duplicated(names(gene_list)) )  ) {\n",
    "    warning(\"Duplicates in gene names\")\n",
    "    gene_list = gene_list[!duplicated(names(gene_list))]\n",
    "  }\n",
    "  if  ( !all( order(gene_list, decreasing = TRUE) == 1:length(gene_list)) ){\n",
    "    warning(\"Gene list not sorted\")\n",
    "    gene_list = sort(gene_list, decreasing = TRUE)\n",
    "  }\n",
    "  myGO = fgsea::gmtPathways(GO_file)\n",
    "\n",
    "  fgRes <- fgsea::fgsea(pathways = myGO,\n",
    "                           stats = gene_list,\n",
    "                           minSize=15, ## minimum gene set size\n",
    "                           maxSize=400, ## maximum gene set size\n",
    "                           nperm=10000) %>% \n",
    "                  as.data.frame() %>% \n",
    "                  dplyr::filter(padj < !!pval) %>% \n",
    "                  arrange(desc(NES))\n",
    "  message(paste(\"Number of signficant gene sets =\", nrow(fgRes)))\n",
    "\n",
    "  message(\"Collapsing Pathways -----\")\n",
    "  concise_pathways = collapsePathways(data.table::as.data.table(fgRes),\n",
    "                                      pathways = myGO,\n",
    "                                      stats = gene_list)\n",
    "  fgRes = fgRes[fgRes$pathway %in% concise_pathways$mainPathways, ]\n",
    "  message(paste(\"Number of gene sets after collapsing =\", nrow(fgRes)))\n",
    "\n",
    "  fgRes$Enrichment = ifelse(fgRes$NES > 0, \"Up-regulated\", \"Down-regulated\")\n",
    "  filtRes = rbind(head(fgRes, n = 10),\n",
    "                  tail(fgRes, n = 10 ))\n",
    "\n",
    "  total_up = sum(fgRes$Enrichment == \"Up-regulated\")\n",
    "  total_down = sum(fgRes$Enrichment == \"Down-regulated\")\n",
    "  header = paste0(\"Top 10 (Total pathways: Up=\", total_up,\", Down=\",    total_down, \")\")\n",
    "\n",
    "  colos = setNames(c(\"firebrick2\", \"dodgerblue2\"),\n",
    "                 c(\"Up-regulated\", \"Down-regulated\"))\n",
    "\n",
    "g1= ggplot(filtRes, aes(reorder(pathway, NES), NES)) +\n",
    "  geom_point( aes(fill = Enrichment, size = size), shape=21) +\n",
    "  scale_fill_manual(values = colos ) +\n",
    "  scale_size_continuous(range = c(2,10)) +\n",
    "  geom_hline(yintercept = 0) +\n",
    "  coord_flip() +\n",
    "  labs(x=\"Pathway\", y=\"Normalized Enrichment Score\",\n",
    "       title=header) + \n",
    "        th\n",
    "\n",
    "  output = list(\"Results\" = fgRes, \"Plot\" = g1)\n",
    "  return(output)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "282fd09a-b14f-4ebd-91de-2e49db9bc357",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>entrez_id</th><th scope=col>logFC</th><th scope=col>pval</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>LOC111099115</td><td>111099115</td><td>1.7885864</td><td>4.161092e-06</td><td>0.0002004183</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>LOC111099216</td><td>111099216</td><td>1.4828614</td><td>7.544561e-04</td><td>0.0125496496</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>LOC111099417</td><td>111099417</td><td>2.7487429</td><td>6.452048e-05</td><td>0.0019567084</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>LOC111099424</td><td>111099424</td><td>3.4384445</td><td>3.211770e-05</td><td>0.0010745946</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>LOC111099722</td><td>111099722</td><td>0.8993515</td><td>1.093514e-04</td><td>0.0028699825</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>LOC111099918</td><td>111099918</td><td>0.6567512</td><td>3.539745e-03</td><td>0.0407210025</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & gene & entrez\\_id & logFC & pval & padj\\\\\n",
       "  & <chr> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & LOC111099115 & 111099115 & 1.7885864 & 4.161092e-06 & 0.0002004183\\\\\n",
       "\t2 & LOC111099216 & 111099216 & 1.4828614 & 7.544561e-04 & 0.0125496496\\\\\n",
       "\t3 & LOC111099417 & 111099417 & 2.7487429 & 6.452048e-05 & 0.0019567084\\\\\n",
       "\t4 & LOC111099424 & 111099424 & 3.4384445 & 3.211770e-05 & 0.0010745946\\\\\n",
       "\t5 & LOC111099722 & 111099722 & 0.8993515 & 1.093514e-04 & 0.0028699825\\\\\n",
       "\t6 & LOC111099918 & 111099918 & 0.6567512 & 3.539745e-03 & 0.0407210025\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | entrez_id &lt;int&gt; | logFC &lt;dbl&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | LOC111099115 | 111099115 | 1.7885864 | 4.161092e-06 | 0.0002004183 |\n",
       "| 2 | LOC111099216 | 111099216 | 1.4828614 | 7.544561e-04 | 0.0125496496 |\n",
       "| 3 | LOC111099417 | 111099417 | 2.7487429 | 6.452048e-05 | 0.0019567084 |\n",
       "| 4 | LOC111099424 | 111099424 | 3.4384445 | 3.211770e-05 | 0.0010745946 |\n",
       "| 5 | LOC111099722 | 111099722 | 0.8993515 | 1.093514e-04 | 0.0028699825 |\n",
       "| 6 | LOC111099918 | 111099918 | 0.6567512 | 3.539745e-03 | 0.0407210025 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene         entrez_id logFC     pval         padj        \n",
       "1 LOC111099115 111099115 1.7885864 4.161092e-06 0.0002004183\n",
       "2 LOC111099216 111099216 1.4828614 7.544561e-04 0.0125496496\n",
       "3 LOC111099417 111099417 2.7487429 6.452048e-05 0.0019567084\n",
       "4 LOC111099424 111099424 3.4384445 3.211770e-05 0.0010745946\n",
       "5 LOC111099722 111099722 0.8993515 1.093514e-04 0.0028699825\n",
       "6 LOC111099918 111099918 0.6567512 3.539745e-03 0.0407210025"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(Entrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "f91e9481-c523-4770-9464-17191163e265",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>111114820</dt><dd>7.30455059714696</dd><dt>111114521</dt><dd>6.53848195612027</dd><dt>111119752</dt><dd>6.24813789296976</dd><dt>111107999</dt><dd>6.05372348435752</dd><dt>111105528</dt><dd>6.01734437817952</dd><dt>111125118</dt><dd>5.96543381117296</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[111114820] 7.30455059714696\n",
       "\\item[111114521] 6.53848195612027\n",
       "\\item[111119752] 6.24813789296976\n",
       "\\item[111107999] 6.05372348435752\n",
       "\\item[111105528] 6.01734437817952\n",
       "\\item[111125118] 5.96543381117296\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "111114820\n",
       ":   7.30455059714696111114521\n",
       ":   6.53848195612027111119752\n",
       ":   6.24813789296976111107999\n",
       ":   6.05372348435752111105528\n",
       ":   6.01734437817952111125118\n",
       ":   5.96543381117296\n",
       "\n"
      ],
      "text/plain": [
       "111114820 111114521 111119752 111107999 111105528 111125118 \n",
       " 7.304551  6.538482  6.248138  6.053723  6.017344  5.965434 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(dplyr)\n",
    "\n",
    "sig_genes = read.csv(\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/significant_genes/sig_p1_wc_genes.csv\", sep='\\t')\n",
    "gene_list = Entrez$logFC\n",
    "names(gene_list) = Entrez$entrez_id\n",
    "gene_list = sort(gene_list, decreasing = TRUE)\n",
    "gene_list = gene_list[!duplicated(names(gene_list))]\n",
    "head(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "abfcf4ae-8f24-4fe7-a4c0-b281cce74076",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>GO_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>LOC111133408</td><td>GO:2001070           </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>LOC111121603</td><td>GO:2000781,GO:2000781</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>LOC111132389</td><td>GO:2000145           </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>LOC111115105</td><td>GO:1990904,GO:1990904</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>LOC111129853</td><td>GO:1990904,GO:1990904</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>LOC111101512</td><td>GO:1990904           </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & gene & GO\\_id\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & LOC111133408 & GO:2001070           \\\\\n",
       "\t2 & LOC111121603 & GO:2000781,GO:2000781\\\\\n",
       "\t3 & LOC111132389 & GO:2000145           \\\\\n",
       "\t4 & LOC111115105 & GO:1990904,GO:1990904\\\\\n",
       "\t5 & LOC111129853 & GO:1990904,GO:1990904\\\\\n",
       "\t6 & LOC111101512 & GO:1990904           \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | GO_id &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | LOC111133408 | GO:2001070            |\n",
       "| 2 | LOC111121603 | GO:2000781,GO:2000781 |\n",
       "| 3 | LOC111132389 | GO:2000145            |\n",
       "| 4 | LOC111115105 | GO:1990904,GO:1990904 |\n",
       "| 5 | LOC111129853 | GO:1990904,GO:1990904 |\n",
       "| 6 | LOC111101512 | GO:1990904            |\n",
       "\n"
      ],
      "text/plain": [
       "  gene         GO_id                \n",
       "1 LOC111133408 GO:2001070           \n",
       "2 LOC111121603 GO:2000781,GO:2000781\n",
       "3 LOC111132389 GO:2000145           \n",
       "4 LOC111115105 GO:1990904,GO:1990904\n",
       "5 LOC111129853 GO:1990904,GO:1990904\n",
       "6 LOC111101512 GO:1990904           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "GO_file <- read.csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/GO_enrichment_analysis/geneID2GO.txt', row.names=NULL, sep='\\t')\n",
    "colnames(GO_file) = c('gene', 'GO_id')\n",
    "head(GO_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "25261f7a-0b74-474b-926a-c1e42788f530",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>entrez_id</th><th scope=col>GO_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>70</th><td>111099115</td><td>GO:0016255, GO:0042765                                                                                                                                                                                                                                                                   </td></tr>\n",
       "\t<tr><th scope=row>154</th><td>111099424</td><td>GO:0005794                                                                                                                                                                                                                                                                               </td></tr>\n",
       "\t<tr><th scope=row>323</th><td>111099722</td><td>GO:0016020,GO:0016020                                                                                                                                                                                                                                                                    </td></tr>\n",
       "\t<tr><th scope=row>514</th><td>111100092</td><td>GO:0004672, GO:0005524                                                                                                                                                                                                                                                                   </td></tr>\n",
       "\t<tr><th scope=row>536</th><td>111100143</td><td>GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264                                                                                                                                                                                              </td></tr>\n",
       "\t<tr><th scope=row>537</th><td>111100148</td><td>GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & entrez\\_id & GO\\_id\\\\\n",
       "  & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t70 & 111099115 & GO:0016255, GO:0042765                                                                                                                                                                                                                                                                   \\\\\n",
       "\t154 & 111099424 & GO:0005794                                                                                                                                                                                                                                                                               \\\\\n",
       "\t323 & 111099722 & GO:0016020,GO:0016020                                                                                                                                                                                                                                                                    \\\\\n",
       "\t514 & 111100092 & GO:0004672, GO:0005524                                                                                                                                                                                                                                                                   \\\\\n",
       "\t536 & 111100143 & GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264                                                                                                                                                                                              \\\\\n",
       "\t537 & 111100148 & GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | entrez_id &lt;int&gt; | GO_id &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 70 | 111099115 | GO:0016255, GO:0042765                                                                                                                                                                                                                                                                    |\n",
       "| 154 | 111099424 | GO:0005794                                                                                                                                                                                                                                                                                |\n",
       "| 323 | 111099722 | GO:0016020,GO:0016020                                                                                                                                                                                                                                                                     |\n",
       "| 514 | 111100092 | GO:0004672, GO:0005524                                                                                                                                                                                                                                                                    |\n",
       "| 536 | 111100143 | GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264                                                                                                                                                                                               |\n",
       "| 537 | 111100148 | GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042 |\n",
       "\n"
      ],
      "text/plain": [
       "    entrez_id\n",
       "70  111099115\n",
       "154 111099424\n",
       "323 111099722\n",
       "514 111100092\n",
       "536 111100143\n",
       "537 111100148\n",
       "    GO_id                                                                                                                                                                                                                                                                                    \n",
       "70  GO:0016255, GO:0042765                                                                                                                                                                                                                                                                   \n",
       "154 GO:0005794                                                                                                                                                                                                                                                                               \n",
       "323 GO:0016020,GO:0016020                                                                                                                                                                                                                                                                    \n",
       "514 GO:0004672, GO:0005524                                                                                                                                                                                                                                                                   \n",
       "536 GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264,GO:0005085, GO:0007264                                                                                                                                                                                              \n",
       "537 GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042,GO:0004435, GO:0005085, GO:0007264, GO:0016042"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "merge_df <- na.omit(merge(convert, GO_file, by='gene', all=TRUE))\n",
    "GO_file <- select(merge_df, entrez_id, GO_id)\n",
    "\n",
    "head(GO_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a6ce59ae-6f33-4baf-be9e-7a3df0476441",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>111114820</dt><dd>7.30455059714696</dd><dt>111114521</dt><dd>6.53848195612027</dd><dt>111119752</dt><dd>6.24813789296976</dd><dt>111107999</dt><dd>6.05372348435752</dd><dt>111105528</dt><dd>6.01734437817952</dd><dt>111125118</dt><dd>5.96543381117296</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[111114820] 7.30455059714696\n",
       "\\item[111114521] 6.53848195612027\n",
       "\\item[111119752] 6.24813789296976\n",
       "\\item[111107999] 6.05372348435752\n",
       "\\item[111105528] 6.01734437817952\n",
       "\\item[111125118] 5.96543381117296\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "111114820\n",
       ":   7.30455059714696111114521\n",
       ":   6.53848195612027111119752\n",
       ":   6.24813789296976111107999\n",
       ":   6.05372348435752111105528\n",
       ":   6.01734437817952111125118\n",
       ":   5.96543381117296\n",
       "\n"
      ],
      "text/plain": [
       "111114820 111114521 111119752 111107999 111105528 111125118 \n",
       " 7.304551  6.538482  6.248138  6.053723  6.017344  5.965434 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c85e8ab3-08e0-410c-abde-6afcfa825c88",
   "metadata": {},
   "source": [
    "need to look into expected input format of these variables..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "c76eff82-e696-4b79-9293-b712d5b4a689",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in readLines(gmt.file): 'con' is not a connection\n",
     "output_type": "error",
     "traceback": [
      "Error in readLines(gmt.file): 'con' is not a connection\nTraceback:\n",
      "1. GSEA(gene_list, GO_file, pval = 0.05)",
      "2. fgsea::gmtPathways(GO_file)   # at line 14 of file <text>",
      "3. strsplit(readLines(gmt.file), \"\\t\")",
      "4. readLines(gmt.file)"
     ]
    }
   ],
   "source": [
    "res = GSEA(gene_list, GO_file, pval = 0.05)\n",
    "dim(res$Results)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:.conda-r-env]",
   "language": "R",
   "name": "conda-env-.conda-r-env-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
