{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a42f122a-183c-48fc-a0ac-0662181fe513",
   "metadata": {},
   "source": [
    "# overlap of methylated CpG with gene features\n",
    "\n",
    "Would like to know how many CpG dinucleotides overlap with genomic features, as well as methylated CpGs\n",
    "\n",
    "the genomic feature files below were generated in [generate_genomic_feature_tracks.ipynb](https://github.com/jgmcdonough/CE18_methylRAD_analysis/blob/master/analysis/genomic_feature_tracks/generate_genomic_feature_tracks.ipynb)\n",
    "\n",
    "using the [BEDtools suite](https://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "092fc196-ffb7-45d2-96d1-b9e86101b008",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene feature BED files\n",
    "exonList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/CV_exon_sorted.bed\"\n",
    "intronList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/CV_intron.bed\"\n",
    "exonUTR=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/CV_exonUTR.bed\"\n",
    "promoterList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/mRNA_promoter_track.bed\"\n",
    "intergenicList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/CV_intergenic.bed\"\n",
    "teList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/genomic_feature_tracks/Venkataraman_files/C_virginica-3.0_TE-all.gff\"\n",
    "geneList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/CV_gene_sorted.bed\"\n",
    "noncodingList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/CV_noncoding.bed\"\n",
    "cdsList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/CV_CDS_sorted.bed\"\n",
    "\n",
    "# CpG dinucleotide list\n",
    "cpgList=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/C_virginica-3.0_CG-motif.bed\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19c19e9e-b727-4f9e-9a0c-a8ac60e797cf",
   "metadata": {},
   "source": [
    "**all CpG dinucleotides** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "95d28e4f-1166-4934-a361-0bf1d1b7b536",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7774010\n",
      "total CpG dinucleotides in genes\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {geneList} | wc -l\n",
    "!echo \"total CpG dinucleotides in genes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c92b89f4-3276-47e3-8ba0-18885475e69c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2323389\n",
      "total CpG dinucleotides in exons\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {exonList} | wc -l\n",
    "!echo \"total CpG dinucleotides in exons\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9913fafc-143f-4439-a9bf-38f4ce43513c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5497874\n",
      "total CpG dinucleotides in introns\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {intronList} | wc -l\n",
    "!echo \"total CpG dinucleotides in introns\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1a78c48b-8dd8-4d07-a559-dc8326ed15b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "600840\n",
      "total CpG dinucleotides in UTRs\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {exonUTR} | wc -l\n",
    "!echo \"total CpG dinucleotides in UTRs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3ed0bf70-e0f1-4ba4-8f00-fec4d6e210eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "926518\n",
      "total CpG dinucleotides in putative promoters\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {promoterList} | wc -l\n",
    "!echo \"total CpG dinucleotides in putative promoters\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "97c5e38c-9cfe-4cd9-97a1-18586a55bb78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6644297\n",
      "total CpG dinucleotides in intergenic regions\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {intergenicList} | wc -l\n",
    "!echo \"total CpG dinucleotides in intergenic regions\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f1adeee1-ff43-40c2-a58d-b5c45b26f152",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12142171\n",
      "total CpG dinucleotides in non-coding regions\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {noncodingList} | wc -l\n",
    "!echo \"total CpG dinucleotides in non-coding regions\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "92558bec-c6d9-4999-bffb-e0f207d0fc44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2828372\n",
      "total CpG dinucleotides in transposable elements\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {teList} | wc -l\n",
    "!echo \"total CpG dinucleotides in transposable elements\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0d3985c5-8683-49aa-bb5d-da66891009cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1722555\n",
      "total CpG dinucleotides in coding sequences\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {cpgList} -b {cdsList} | wc -l\n",
    "!echo \"total CpG dinucleotides in coding sequences\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c17cf1de-5c9a-43d6-8d02-31e621ce1dac",
   "metadata": {},
   "source": [
    "#### methylated CpG by treatment\n",
    "\n",
    "first, filtering treatment CpG motifs - a CpG dinucleotide is considered methylated if the average count across the four replicates is greater than or equal to 4. This ensures that we're not counting a CpG methylated if one sample has a count of 15 and the rest have 0.\n",
    "\n",
    "only need to run this code once - then can just read in files below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "23ed9fb0-3312-4a4f-a2b3-74be81432781",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV file\n",
    "CC_multicov = pd.read_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CC_CpGmulticov.csv')\n",
    "\n",
    "# Calculate average counts of last 5 columns\n",
    "CC_multicov['avg_counts'] = CC_multicov.iloc[:, -5:].mean(axis=1)\n",
    "\n",
    "# Remove rows with averages below 5\n",
    "CC_filtered = CC_multicov[CC_multicov['avg_counts'] >= 5]\n",
    "\n",
    "# Remove temporary average column\n",
    "CC_filtered = CC_filtered.drop('avg_counts', axis=1)\n",
    "\n",
    "# Select the desired columns\n",
    "CC_bed = CC_filtered[['chromosome', 'start', 'stop']]\n",
    "\n",
    "# Save to bed file\n",
    "CC_bed.to_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CC_cpgMethyl.bed', \n",
    "               sep='\\t', \n",
    "               header=False, \n",
    "               index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "91e1e2b9-ecdf-4341-9704-0184dd82c04f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the CSV file\n",
    "HH_multicov = pd.read_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_CpGmulticov.csv')\n",
    "\n",
    "# Calculate average counts of last 5 columns\n",
    "HH_multicov['avg_counts'] = HH_multicov.iloc[:, -5:].mean(axis=1)\n",
    "\n",
    "# Remove rows with averages below 5\n",
    "HH_filtered = HH_multicov[HH_multicov['avg_counts'] >= 5]\n",
    "\n",
    "# Remove temporary average column\n",
    "HH_filtered = HH_filtered.drop('avg_counts', axis=1)\n",
    "\n",
    "# Select the desired columns\n",
    "HH_bed = HH_filtered[['chromosome', 'start', 'stop']]\n",
    "\n",
    "# Save to bed file\n",
    "HH_bed.to_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_cpgMethyl.bed', \n",
    "               sep='\\t', \n",
    "               header=False, \n",
    "               index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "9922acb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the CSV file\n",
    "HH_multicov = pd.read_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_CpGmulticov.csv')\n",
    "\n",
    "# Calculate average counts of last 5 columns\n",
    "HH_multicov['avg_counts'] = HH_multicov.iloc[:, -5:].mean(axis=1)\n",
    "\n",
    "# Remove rows with averages below 5\n",
    "HH_filtered = HH_multicov[HH_multicov['avg_counts'] >= 5]\n",
    "\n",
    "# Remove temporary average column\n",
    "HH_filtered = HH_filtered.drop('avg_counts', axis=1)\n",
    "\n",
    "# Select the desired columns\n",
    "HH_bed = HH_filtered[['chromosome', 'start', 'stop']]\n",
    "\n",
    "# Save to bed file\n",
    "HH_bed.to_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_cpgMethyl.bed', \n",
    "               sep='\\t', \n",
    "               header=False, \n",
    "               index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "d69e41b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the CSV file\n",
    "HH_multicov = pd.read_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_CpGmulticov.csv')\n",
    "\n",
    "# Calculate average counts of last 5 columns\n",
    "HH_multicov['avg_counts'] = HH_multicov.iloc[:, -5:].mean(axis=1)\n",
    "\n",
    "# Remove rows with averages below 5\n",
    "HH_filtered = HH_multicov[HH_multicov['avg_counts'] >= 5]\n",
    "\n",
    "# Remove temporary average column\n",
    "HH_filtered = HH_filtered.drop('avg_counts', axis=1)\n",
    "\n",
    "# Select the desired columns\n",
    "HH_bed = HH_filtered[['chromosome', 'start', 'stop']]\n",
    "\n",
    "# Save to bed file\n",
    "HH_bed.to_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_cpgMethyl.bed', \n",
    "               sep='\\t', \n",
    "               header=False, \n",
    "               index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "66d7a88a-8f17-40d6-af94-5ff761977703",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CpG lists for each treatment\n",
    "CC_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CC_cpgMethyl.bed\"\n",
    "CH_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CH_cpgMethyl.bed\"\n",
    "HC_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HC_cpgMethyl.bed\"\n",
    "HH_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_cpgMethyl.bed\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "687fdff9-cffa-4fb2-a11c-a3c2ec41f887",
   "metadata": {},
   "source": [
    "**methylated CpG in genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "e25e1aee-4cfd-43c1-8aae-c390066d29c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86816\n",
      "methylated CpG for CC overlaps with genes\n",
      "37891\n",
      "methylated CpG for CH overlaps with genes\n",
      "110613\n",
      "methylated CpG for HC overlaps with genes\n",
      "101771\n",
      "methylated CpG for HH overlaps with genes\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with genes\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with genes\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with genes\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with genes\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dad89b81-6f3a-46c2-a553-77791cd83f6d",
   "metadata": {},
   "source": [
    "**methylated CpG in exons**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "505aeef2-9868-45e3-ab2b-f9c3975b50dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57957\n",
      "methylated CpG for CC overlaps with exons\n",
      "26233\n",
      "methylated CpG for CH overlaps with exons\n",
      "68256\n",
      "methylated CpG for HC overlaps with exons\n",
      "64813\n",
      "methylated CpG for HH overlaps with exons\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with exons\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with exons\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with exons\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with exons\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8f02a76-67cb-44a7-bf07-f2364e0baf25",
   "metadata": {},
   "source": [
    "**methylated CpG in introns**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "9fb39afb-3155-4b30-9e89-ef7986332fef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29638\n",
      "methylated CpG for CC overlaps with introns\n",
      "12045\n",
      "methylated CpG for CH overlaps with introns\n",
      "43310\n",
      "methylated CpG for HC overlaps with introns\n",
      "37891\n",
      "methylated CpG for HH overlaps with introns\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with introns\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with introns\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with introns\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with introns\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "078139b9-e98e-4864-9045-fe4a72e18a20",
   "metadata": {},
   "source": [
    "**methylated CpG in putative promoters**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "0d85424c-8b69-4b87-8606-6d9a437e16ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3529\n",
      "methylated CpG for CC overlaps with promoters\n",
      "1606\n",
      "methylated CpG for CH overlaps with promoters\n",
      "5650\n",
      "methylated CpG for HC overlaps with promoters\n",
      "4889\n",
      "methylated CpG for HH overlaps with promoters\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with promoters\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eba5bdce-6ab8-4c7f-864b-6eced5bb5db8",
   "metadata": {},
   "source": [
    "**methylated CpG in exon UTRs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "29d859bb-41de-4277-9527-c2a2c839b1ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4165\n",
      "methylated CpG for CC overlaps with exon UTRs\n",
      "1545\n",
      "methylated CpG for CH overlaps with exon UTRs\n",
      "5164\n",
      "methylated CpG for HC overlaps with exon UTRs\n",
      "4843\n",
      "methylated CpG for HH overlaps with exon UTRs\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with exon UTRs\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with exon UTRs\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with exon UTRs\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with exon UTRs\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8fa8d21-772f-4732-bb99-e3afa6b63ee9",
   "metadata": {},
   "source": [
    "**methylated CpG in transposable elements**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "b9c38fb1-163b-4af2-bfb9-0a421fadf608",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16875\n",
      "methylated CpG for CC overlaps with TEs\n",
      "6346\n",
      "methylated CpG for CH overlaps with TEs\n",
      "24506\n",
      "methylated CpG for HC overlaps with TEs\n",
      "22736\n",
      "methylated CpG for HH overlaps with TEs\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with TEs\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with TEs\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with TEs\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with TEs\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be218760-8594-4d79-bd89-842f1c4650fc",
   "metadata": {},
   "source": [
    "**methylated CpG in intergenic regions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "efd07ec5-225d-4f0d-ac8a-e81330c0c4c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20079\n",
      "methylated CpG for CC overlaps with intergenic regions\n",
      "6567\n",
      "methylated CpG for CH overlaps with intergenic regions\n",
      "32953\n",
      "methylated CpG for HC overlaps with intergenic regions\n",
      "28033\n",
      "methylated CpG for HH overlaps with intergenic regions\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with intergenic regions\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with intergenic regions\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with intergenic regions\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with intergenic regions\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cc96763-bc79-455c-a8ce-9a051a95c580",
   "metadata": {},
   "source": [
    "## no overlaps with genomic features\n",
    "\n",
    "CpGs that do not overlap any feature (aka unannotated intergenic regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "a8ca491a-bc77-4d86-b6fe-563a9252799c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4576705\n",
      "CpG motifs do not overlap with exons, introns, transposable elements (all), or putative promoters\n"
     ]
    }
   ],
   "source": [
    "# CpG motif\n",
    "!bedtools intersect -v -a {cpgList} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"CpG motifs do not overlap with exons, introns, transposable elements (all), or putative promoters\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566bf46b-a50c-4528-bbf1-5f1cc05dcb6a",
   "metadata": {},
   "source": [
    "4499027 CpG motifs do not overlap with exons, introns, transposable elements (all), or putative promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d41b6582",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12610\n",
      "methylated CpG for cont cont do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n",
      "3837\n",
      "methylated CpG for cont hyp do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n",
      "20716\n",
      "methylated CpG for hyp cont do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n",
      "16967\n",
      "methylated CpG for hyp hyp do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -v -a {CC_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"methylated CpG for cont cont do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\"\n",
    "\n",
    "!bedtools intersect -v -a {CH_CpG} -b {exonList} {intronList} {teList} {promoterList}  | wc -l\n",
    "!echo \"methylated CpG for cont hyp do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\"\n",
    "\n",
    "!bedtools intersect -v -a {HC_CpG} -b {exonList} {intronList} {teList} {promoterList}  | wc -l\n",
    "!echo \"methylated CpG for hyp cont do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\"\n",
    "\n",
    "!bedtools intersect -v -a {HH_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"methylated CpG for hyp hyp do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9a54378-2d81-4cf7-9813-a0181a8bfd1d",
   "metadata": {},
   "source": [
    "**how many methylated CpG overlap outside of genes? (not in genic regions)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "127d5d41-cf0d-4715-b597-97617a7fcc0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20827\n",
      "methylated CpG for cont cont do not overlap with genes\n",
      "6934\n",
      "methylated CpG for cont hyp do not overlap with genes\n",
      "33868\n",
      "methylated CpG for hyp cont do not overlap with genes\n",
      "28935\n",
      "methylated CpG for hyp hyp do not overlap with genes\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -v -a {CC_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for cont cont do not overlap with genes\"\n",
    "\n",
    "!bedtools intersect -v -a {CH_CpG} -b {geneList}  | wc -l\n",
    "!echo \"methylated CpG for cont hyp do not overlap with genes\"\n",
    "\n",
    "!bedtools intersect -v -a {HC_CpG} -b {geneList}  | wc -l\n",
    "!echo \"methylated CpG for hyp cont do not overlap with genes\"\n",
    "\n",
    "!bedtools intersect -v -a {HH_CpG} -b {geneList}  | wc -l\n",
    "!echo \"methylated CpG for hyp hyp do not overlap with genes\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "988d513a-8401-4725-a858-3f2a40f2f383",
   "metadata": {},
   "source": [
    "## Proportion overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "969d71a4-222d-4cff-9d5e-6517ac743664",
   "metadata": {},
   "outputs": [],
   "source": [
    "CC_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CC_cpgMethyl.bed\"\n",
    "CH_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CH_cpgMethyl.bed\"\n",
    "HC_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HC_cpgMethyl.bed\"\n",
    "HH_CpG = \"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_cpgMethyl.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "329ce885-e237-4758-be69-fee4a4f448df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      genomicFeature treatment propOverlap\n",
      "0                 TE        CC     [16875]\n",
      "1              exons        CC     [57957]\n",
      "2            introns        CC     [29638]\n",
      "3   putativePromoter        CC      [3529]\n",
      "4               UTRs        CC      [4165]\n",
      "5         intergenic        CC     [20079]\n",
      "6                 TE        CH      [6346]\n",
      "7              exons        CH     [26233]\n",
      "8            introns        CH     [12045]\n",
      "9   putativePromoter        CH      [1606]\n",
      "10              UTRs        CH      [1545]\n",
      "11        intergenic        CH      [6567]\n",
      "12                TE        HC     [24506]\n",
      "13             exons        HC     [68256]\n",
      "14           introns        HC     [43310]\n",
      "15  putativePromoter        HC      [5650]\n",
      "16              UTRs        HC      [5164]\n",
      "17        intergenic        HC     [32953]\n",
      "18                TE        HH     [22736]\n",
      "19             exons        HH     [64813]\n",
      "20           introns        HH     [37891]\n",
      "21  putativePromoter        HH      [4889]\n",
      "22              UTRs        HH      [4843]\n",
      "23        intergenic        HH     [28033]\n",
      "24                TE    allCpG   [2828372]\n",
      "25             exons    allCpG   [2323389]\n",
      "26           introns    allCpG   [5497874]\n",
      "27  putativePromoter    allCpG    [926518]\n",
      "28              UTRs    allCpG    [600840]\n",
      "29        intergenic    allCpG   [6644297]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define the treatments and feature variables\n",
    "treatments = {\n",
    "    'CC': CC_CpG,\n",
    "    'CH': CH_CpG,\n",
    "    'HC': HC_CpG, \n",
    "    'HH': HH_CpG,\n",
    "    'allCpG': cpgList\n",
    "}\n",
    "\n",
    "features = {\n",
    "    'TE': teList,\n",
    "    'exons': exonList,\n",
    "    'introns': intronList,\n",
    "    'putativePromoter': promoterList,\n",
    "    'UTRs': exonUTR,\n",
    "    'intergenic': intergenicList\n",
    "}\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "results = []\n",
    "\n",
    "# Loop over the treatments\n",
    "for treatment_name, treatment_file in treatments.items():\n",
    "    # Loop over the features\n",
    "    for feature_name, feature_list in features.items():\n",
    "        # Use bedtools intersect to count the overlaps\n",
    "        overlap_count = !bedtools intersect -u -a {treatment_file} -b {feature_list} | wc -l\n",
    "        \n",
    "        # Append the result to the list\n",
    "        results.append({'genomicFeature': feature_name, 'treatment': treatment_name, 'propOverlap': overlap_count})\n",
    "\n",
    "\n",
    "# Convert the list to a pandas DataFrame\n",
    "df = pd.DataFrame(results)\n",
    "\n",
    "# Print the DataFrame\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "135db239-b750-4285-ad72-db9f773ed62b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['propOverlap'] = df['propOverlap'].str.extract('(\\d+)')\n",
    "\n",
    "df.to_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/propOverlap.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12e753f5-c31d-440a-9254-94126ab383ff",
   "metadata": {},
   "source": [
    "can't seem to get the propOverlap column formatted correctly, so manipulated locally and then re-uploaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "d7884d06-4721-4d53-b9b4-c287e8c0f4a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      genomicFeature treatment  propOverlap  propOverlap_sum  \\\n",
      "0                 TE        CC        16875           132243   \n",
      "1              exons        CC        57957           132243   \n",
      "2            introns        CC        29638           132243   \n",
      "3   putativePromoter        CC         3529           132243   \n",
      "4               UTRs        CC         4165           132243   \n",
      "5         intergenic        CC        20079           132243   \n",
      "6                 TE        CH         6346            54342   \n",
      "7              exons        CH        26233            54342   \n",
      "8            introns        CH        12045            54342   \n",
      "9   putativePromoter        CH         1606            54342   \n",
      "10              UTRs        CH         1545            54342   \n",
      "11        intergenic        CH         6567            54342   \n",
      "12                TE        HC        24506           179839   \n",
      "13             exons        HC        68256           179839   \n",
      "14           introns        HC        43310           179839   \n",
      "15  putativePromoter        HC         5650           179839   \n",
      "16              UTRs        HC         5164           179839   \n",
      "17        intergenic        HC        32953           179839   \n",
      "18                TE        HH        22736           163205   \n",
      "19             exons        HH        64813           163205   \n",
      "20           introns        HH        37891           163205   \n",
      "21  putativePromoter        HH         4889           163205   \n",
      "22              UTRs        HH         4843           163205   \n",
      "23        intergenic        HH        28033           163205   \n",
      "24                TE    allCpG      2828372         18821290   \n",
      "25             exons    allCpG      2323389         18821290   \n",
      "26           introns    allCpG      5497874         18821290   \n",
      "27  putativePromoter    allCpG       926518         18821290   \n",
      "28              UTRs    allCpG       600840         18821290   \n",
      "29        intergenic    allCpG      6644297         18821290   \n",
      "\n",
      "    proportion_overlap  \n",
      "0             0.127606  \n",
      "1             0.438261  \n",
      "2             0.224118  \n",
      "3             0.026686  \n",
      "4             0.031495  \n",
      "5             0.151834  \n",
      "6             0.116779  \n",
      "7             0.482739  \n",
      "8             0.221652  \n",
      "9             0.029554  \n",
      "10            0.028431  \n",
      "11            0.120846  \n",
      "12            0.136266  \n",
      "13            0.379539  \n",
      "14            0.240827  \n",
      "15            0.031417  \n",
      "16            0.028715  \n",
      "17            0.183236  \n",
      "18            0.139309  \n",
      "19            0.397126  \n",
      "20            0.232168  \n",
      "21            0.029956  \n",
      "22            0.029674  \n",
      "23            0.171766  \n",
      "24            0.150275  \n",
      "25            0.123445  \n",
      "26            0.292109  \n",
      "27            0.049227  \n",
      "28            0.031923  \n",
      "29            0.353020  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load your data into a Pandas DataFrame\n",
    "df = pd.read_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/propOverlap.csv')\n",
    "\n",
    "# Group the data by treatment and calculate the sum of propOverlap\n",
    "treatment_sums = df.groupby('treatment')['propOverlap'].sum().reset_index()\n",
    "\n",
    "# Merge the original DataFrame with the treatment sums\n",
    "df_merged = pd.merge(df, treatment_sums, on='treatment', suffixes=('', '_sum'))\n",
    "\n",
    "# Calculate the proportion of overlap\n",
    "df_merged['proportion_overlap'] = df_merged['propOverlap'] / df_merged['propOverlap_sum']\n",
    "\n",
    "# Print the result\n",
    "print(df_merged)\n",
    "\n",
    "df_merged.to_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/propOverlap_decimal.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dbfdca9-2fba-466b-bc92-c84bb650ae3a",
   "metadata": {},
   "source": [
    "## Proportion overlap with methylated CpGs\n",
    "\n",
    "There are 5 individual replicate oysters per treatment combination. A CpG dinucleotide is considered methylated for that treatment if the majority of the replicates (at least 3 individuals out of the 5) have >= 5 sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7316217a-11b9-4d4b-8a99-6f6ce341f3cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# List of input CSV files\n",
    "input_files = [\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CC_CpGmulticov.csv',\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/CH_CpGmulticov.csv',\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HC_CpGmulticov.csv',\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/HH_CpGmulticov.csv'\n",
    "]\n",
    "\n",
    "# List of output BED files\n",
    "output_files = [\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_CC_cpgMethyl.bed',\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_CH_cpgMethyl.bed',\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_HC_cpgMethyl.bed',\n",
    "    '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_HH_cpgMethyl.bed'\n",
    "]\n",
    "\n",
    "# Loop through input and output files\n",
    "for input_file, output_file in zip(input_files, output_files):\n",
    "    # Load the CSV file\n",
    "    multicov_file = pd.read_csv(input_file)\n",
    "\n",
    "    # Filter rows with at least 3 columns having 5 or more sequences\n",
    "    filtered_multi = multicov_file[(multicov_file.iloc[:, -5:] >= 5).sum(axis=1) >= 3]\n",
    "\n",
    "    # Select the desired columns\n",
    "    bed_file = filtered_multi[['chromosome', 'start', 'stop']]\n",
    "\n",
    "    # Save to bed file\n",
    "    bed_file.to_csv(output_file, sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d72e4a81-b63c-4597-87af-e6435ce6fe70",
   "metadata": {},
   "outputs": [],
   "source": [
    "CC_CpG = '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_CC_cpgMethyl.bed'\n",
    "CH_CpG = '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_CH_cpgMethyl.bed'\n",
    "HC_CpG = '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_HC_cpgMethyl.bed'\n",
    "HH_CpG = '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_methyl_assembly/assembly_pipeline_files/genomic_bed_files/CpG_multicov/PO_HH_cpgMethyl.bed'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9abc546-7ada-4f21-8ba3-0910cf3ee57e",
   "metadata": {},
   "source": [
    "**methylated CpG in genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c51919e6-2006-480b-b425-8c8271feaa3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61389\n",
      "methylated CpG for CC overlaps with genes\n",
      "68470\n",
      "methylated CpG for CH overlaps with genes\n",
      "91333\n",
      "methylated CpG for HC overlaps with genes\n",
      "86410\n",
      "methylated CpG for HH overlaps with genes\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with genes\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with genes\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with genes\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {geneList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with genes\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef11ac6c-a637-453e-a2a3-1aa6618ea522",
   "metadata": {},
   "source": [
    "**methylated CpG in exons**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b639a3c4-4312-4d7e-980b-981ea2197e2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42577\n",
      "methylated CpG for CC overlaps with exons\n",
      "46527\n",
      "methylated CpG for CH overlaps with exons\n",
      "57498\n",
      "methylated CpG for HC overlaps with exons\n",
      "56100\n",
      "methylated CpG for HH overlaps with exons\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with exons\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with exons\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with exons\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {exonList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with exons\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a361aa3-9ecd-4da1-873a-5097f9de9013",
   "metadata": {},
   "source": [
    "**methylated CpG in introns**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "082b325d-65c7-4c7a-a1d8-6fe7d47c6309",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19342\n",
      "methylated CpG for CC overlaps with introns\n",
      "22545\n",
      "methylated CpG for CH overlaps with introns\n",
      "34622\n",
      "methylated CpG for HC overlaps with introns\n",
      "31070\n",
      "methylated CpG for HH overlaps with introns\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with introns\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with introns\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with introns\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {intronList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with introns\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d7b4e4c-20b1-4337-9709-e4d98a7f2915",
   "metadata": {},
   "source": [
    "**methylated CpG in putative promoters**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e32c255d-c44c-4eed-9c8f-e4a3e6ddd2d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2081\n",
      "methylated CpG for CC overlaps with promoters\n",
      "2712\n",
      "methylated CpG for CH overlaps with promoters\n",
      "4177\n",
      "methylated CpG for HC overlaps with promoters\n",
      "3704\n",
      "methylated CpG for HH overlaps with promoters\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {promoterList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with promoters\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "015d5484-6f7d-490a-a9ab-eaad02912e7b",
   "metadata": {},
   "source": [
    "**methylated CpG in exon UTRs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3c2aa500-68b7-4cc4-957d-e07818af0aa6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2964\n",
      "methylated CpG for CC overlaps with exon UTRs\n",
      "3250\n",
      "methylated CpG for CH overlaps with exon UTRs\n",
      "4207\n",
      "methylated CpG for HC overlaps with exon UTRs\n",
      "3986\n",
      "methylated CpG for HH overlaps with exon UTRs\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with exon UTRs\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with exon UTRs\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with exon UTRs\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {exonUTR} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with exon UTRs\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "428d277e-8bc9-4df1-b57e-fb368b71a457",
   "metadata": {},
   "source": [
    "**methylated CpG in transposable elements**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "60a4207c-e452-414a-a8aa-60d62a422562",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16875\n",
      "methylated CpG for CC overlaps with TEs\n",
      "6346\n",
      "methylated CpG for CH overlaps with TEs\n",
      "24506\n",
      "methylated CpG for HC overlaps with TEs\n",
      "22736\n",
      "methylated CpG for HH overlaps with TEs\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with TEs\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with TEs\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with TEs\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {teList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with TEs\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "124f4e8d-4559-4382-b114-a600bb2b62ff",
   "metadata": {},
   "source": [
    "**methylated CpG in intergenic regions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "786b3ac4-85bb-4565-a79a-3872e8ecdcdb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11929\n",
      "methylated CpG for CC overlaps with intergenic regions\n",
      "15066\n",
      "methylated CpG for CH overlaps with intergenic regions\n",
      "24496\n",
      "methylated CpG for HC overlaps with intergenic regions\n",
      "21279\n",
      "methylated CpG for HH overlaps with intergenic regions\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -u -a {CC_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for CC overlaps with intergenic regions\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for CH overlaps with intergenic regions\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for HC overlaps with intergenic regions\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {intergenicList} | wc -l\n",
    "!echo \"methylated CpG for HH overlaps with intergenic regions\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed784e69-1119-4b20-b9bc-1adc0eb462f2",
   "metadata": {},
   "source": [
    "taking the results above and putting into one df and exporting to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d4f33c60-9fc0-4120-a4e9-8e5af912af37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomicFeature</th>\n",
       "      <th>treatment</th>\n",
       "      <th>propOverlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TE</td>\n",
       "      <td>CC</td>\n",
       "      <td>[11289]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exons</td>\n",
       "      <td>CC</td>\n",
       "      <td>[42577]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>introns</td>\n",
       "      <td>CC</td>\n",
       "      <td>[19342]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CC</td>\n",
       "      <td>[2081]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CC</td>\n",
       "      <td>[2964]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CC</td>\n",
       "      <td>[11929]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TE</td>\n",
       "      <td>CH</td>\n",
       "      <td>[13304]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exons</td>\n",
       "      <td>CH</td>\n",
       "      <td>[46527]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>introns</td>\n",
       "      <td>CH</td>\n",
       "      <td>[22545]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CH</td>\n",
       "      <td>[2712]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CH</td>\n",
       "      <td>[3250]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CH</td>\n",
       "      <td>[15066]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TE</td>\n",
       "      <td>HC</td>\n",
       "      <td>[19594]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>exons</td>\n",
       "      <td>HC</td>\n",
       "      <td>[57498]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>introns</td>\n",
       "      <td>HC</td>\n",
       "      <td>[34622]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HC</td>\n",
       "      <td>[4177]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HC</td>\n",
       "      <td>[4207]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HC</td>\n",
       "      <td>[24496]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TE</td>\n",
       "      <td>HH</td>\n",
       "      <td>[18394]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>exons</td>\n",
       "      <td>HH</td>\n",
       "      <td>[56100]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>introns</td>\n",
       "      <td>HH</td>\n",
       "      <td>[31070]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HH</td>\n",
       "      <td>[3704]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HH</td>\n",
       "      <td>[3986]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HH</td>\n",
       "      <td>[21279]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TE</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>[2828372]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>exons</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>[2323389]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>introns</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>[5497874]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>[926518]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>[600840]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>[6644297]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genomicFeature treatment propOverlap\n",
       "0                 TE        CC     [11289]\n",
       "1              exons        CC     [42577]\n",
       "2            introns        CC     [19342]\n",
       "3   putativePromoter        CC      [2081]\n",
       "4               UTRs        CC      [2964]\n",
       "5         intergenic        CC     [11929]\n",
       "6                 TE        CH     [13304]\n",
       "7              exons        CH     [46527]\n",
       "8            introns        CH     [22545]\n",
       "9   putativePromoter        CH      [2712]\n",
       "10              UTRs        CH      [3250]\n",
       "11        intergenic        CH     [15066]\n",
       "12                TE        HC     [19594]\n",
       "13             exons        HC     [57498]\n",
       "14           introns        HC     [34622]\n",
       "15  putativePromoter        HC      [4177]\n",
       "16              UTRs        HC      [4207]\n",
       "17        intergenic        HC     [24496]\n",
       "18                TE        HH     [18394]\n",
       "19             exons        HH     [56100]\n",
       "20           introns        HH     [31070]\n",
       "21  putativePromoter        HH      [3704]\n",
       "22              UTRs        HH      [3986]\n",
       "23        intergenic        HH     [21279]\n",
       "24                TE    allCpG   [2828372]\n",
       "25             exons    allCpG   [2323389]\n",
       "26           introns    allCpG   [5497874]\n",
       "27  putativePromoter    allCpG    [926518]\n",
       "28              UTRs    allCpG    [600840]\n",
       "29        intergenic    allCpG   [6644297]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define the treatments and feature variables\n",
    "treatments = {\n",
    "    'CC': CC_CpG,\n",
    "    'CH': CH_CpG,\n",
    "    'HC': HC_CpG, \n",
    "    'HH': HH_CpG,\n",
    "    'allCpG': cpgList\n",
    "}\n",
    "\n",
    "features = {\n",
    "    'TE': teList,\n",
    "    'exons': exonList,\n",
    "    'introns': intronList,\n",
    "    'putativePromoter': promoterList,\n",
    "    'UTRs': exonUTR,\n",
    "    'intergenic': intergenicList\n",
    "}\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "results = []\n",
    "\n",
    "# Loop over the treatments\n",
    "for treatment_name, treatment_file in treatments.items():\n",
    "    # Loop over the features\n",
    "    for feature_name, feature_list in features.items():\n",
    "        # Use bedtools intersect to count the overlaps\n",
    "        overlap_count = !bedtools intersect -u -a {treatment_file} -b {feature_list} | wc -l\n",
    "        \n",
    "        # Append the result to the list\n",
    "        results.append({'genomicFeature': feature_name, 'treatment': treatment_name, 'propOverlap': overlap_count})\n",
    "\n",
    "\n",
    "# Convert the list to a pandas DataFrame\n",
    "df = pd.DataFrame(results)\n",
    "\n",
    "# Print the DataFrame\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7b8397fb-0739-46c1-a485-745483d4e357",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomicFeature</th>\n",
       "      <th>treatment</th>\n",
       "      <th>propOverlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TE</td>\n",
       "      <td>CC</td>\n",
       "      <td>11289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exons</td>\n",
       "      <td>CC</td>\n",
       "      <td>42577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>introns</td>\n",
       "      <td>CC</td>\n",
       "      <td>19342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CC</td>\n",
       "      <td>2081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CC</td>\n",
       "      <td>2964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CC</td>\n",
       "      <td>11929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TE</td>\n",
       "      <td>CH</td>\n",
       "      <td>13304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exons</td>\n",
       "      <td>CH</td>\n",
       "      <td>46527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>introns</td>\n",
       "      <td>CH</td>\n",
       "      <td>22545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CH</td>\n",
       "      <td>2712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CH</td>\n",
       "      <td>3250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CH</td>\n",
       "      <td>15066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TE</td>\n",
       "      <td>HC</td>\n",
       "      <td>19594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>exons</td>\n",
       "      <td>HC</td>\n",
       "      <td>57498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>introns</td>\n",
       "      <td>HC</td>\n",
       "      <td>34622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HC</td>\n",
       "      <td>4177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HC</td>\n",
       "      <td>4207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HC</td>\n",
       "      <td>24496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TE</td>\n",
       "      <td>HH</td>\n",
       "      <td>18394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>exons</td>\n",
       "      <td>HH</td>\n",
       "      <td>56100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>introns</td>\n",
       "      <td>HH</td>\n",
       "      <td>31070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HH</td>\n",
       "      <td>3704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HH</td>\n",
       "      <td>3986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HH</td>\n",
       "      <td>21279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TE</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2828372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>exons</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2323389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>introns</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>5497874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>926518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>600840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>6644297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genomicFeature treatment  propOverlap\n",
       "0                 TE        CC        11289\n",
       "1              exons        CC        42577\n",
       "2            introns        CC        19342\n",
       "3   putativePromoter        CC         2081\n",
       "4               UTRs        CC         2964\n",
       "5         intergenic        CC        11929\n",
       "6                 TE        CH        13304\n",
       "7              exons        CH        46527\n",
       "8            introns        CH        22545\n",
       "9   putativePromoter        CH         2712\n",
       "10              UTRs        CH         3250\n",
       "11        intergenic        CH        15066\n",
       "12                TE        HC        19594\n",
       "13             exons        HC        57498\n",
       "14           introns        HC        34622\n",
       "15  putativePromoter        HC         4177\n",
       "16              UTRs        HC         4207\n",
       "17        intergenic        HC        24496\n",
       "18                TE        HH        18394\n",
       "19             exons        HH        56100\n",
       "20           introns        HH        31070\n",
       "21  putativePromoter        HH         3704\n",
       "22              UTRs        HH         3986\n",
       "23        intergenic        HH        21279\n",
       "24                TE    allCpG      2828372\n",
       "25             exons    allCpG      2323389\n",
       "26           introns    allCpG      5497874\n",
       "27  putativePromoter    allCpG       926518\n",
       "28              UTRs    allCpG       600840\n",
       "29        intergenic    allCpG      6644297"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove brackets around the numbers\n",
    "df['propOverlap'] = df['propOverlap'].apply(lambda x: int(x[0]))\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "710299ce-0b06-442f-ab6d-0397db9f60e0",
   "metadata": {},
   "source": [
    "adding in the non-overlap info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "67d5be3b-8009-4f85-893d-7c7ac074b8a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomicFeature</th>\n",
       "      <th>treatment</th>\n",
       "      <th>propOverlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TE</td>\n",
       "      <td>CC</td>\n",
       "      <td>11289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exons</td>\n",
       "      <td>CC</td>\n",
       "      <td>42577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>introns</td>\n",
       "      <td>CC</td>\n",
       "      <td>19342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CC</td>\n",
       "      <td>2081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CC</td>\n",
       "      <td>2964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CC</td>\n",
       "      <td>11929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TE</td>\n",
       "      <td>CH</td>\n",
       "      <td>13304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exons</td>\n",
       "      <td>CH</td>\n",
       "      <td>46527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>introns</td>\n",
       "      <td>CH</td>\n",
       "      <td>22545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CH</td>\n",
       "      <td>2712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CH</td>\n",
       "      <td>3250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CH</td>\n",
       "      <td>15066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TE</td>\n",
       "      <td>HC</td>\n",
       "      <td>19594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>exons</td>\n",
       "      <td>HC</td>\n",
       "      <td>57498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>introns</td>\n",
       "      <td>HC</td>\n",
       "      <td>34622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HC</td>\n",
       "      <td>4177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HC</td>\n",
       "      <td>4207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HC</td>\n",
       "      <td>24496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TE</td>\n",
       "      <td>HH</td>\n",
       "      <td>18394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>exons</td>\n",
       "      <td>HH</td>\n",
       "      <td>56100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>introns</td>\n",
       "      <td>HH</td>\n",
       "      <td>31070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HH</td>\n",
       "      <td>3704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HH</td>\n",
       "      <td>3986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HH</td>\n",
       "      <td>21279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TE</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2828372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>exons</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2323389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>introns</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>5497874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>926518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>600840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>6644297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CC</td>\n",
       "      <td>7345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CH</td>\n",
       "      <td>9137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HC</td>\n",
       "      <td>15273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HH</td>\n",
       "      <td>12963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>4576705</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genomicFeature treatment  propOverlap\n",
       "0                 TE        CC        11289\n",
       "1              exons        CC        42577\n",
       "2            introns        CC        19342\n",
       "3   putativePromoter        CC         2081\n",
       "4               UTRs        CC         2964\n",
       "5         intergenic        CC        11929\n",
       "6                 TE        CH        13304\n",
       "7              exons        CH        46527\n",
       "8            introns        CH        22545\n",
       "9   putativePromoter        CH         2712\n",
       "10              UTRs        CH         3250\n",
       "11        intergenic        CH        15066\n",
       "12                TE        HC        19594\n",
       "13             exons        HC        57498\n",
       "14           introns        HC        34622\n",
       "15  putativePromoter        HC         4177\n",
       "16              UTRs        HC         4207\n",
       "17        intergenic        HC        24496\n",
       "18                TE        HH        18394\n",
       "19             exons        HH        56100\n",
       "20           introns        HH        31070\n",
       "21  putativePromoter        HH         3704\n",
       "22              UTRs        HH         3986\n",
       "23        intergenic        HH        21279\n",
       "24                TE    allCpG      2828372\n",
       "25             exons    allCpG      2323389\n",
       "26           introns    allCpG      5497874\n",
       "27  putativePromoter    allCpG       926518\n",
       "28              UTRs    allCpG       600840\n",
       "29        intergenic    allCpG      6644297\n",
       "30        no_overlap        CC         7345\n",
       "31        no_overlap        CH         9137\n",
       "32        no_overlap        HC        15273\n",
       "33        no_overlap        HH        12963\n",
       "34        no_overlap    allCpG      4576705"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "# Create a dictionary with treatment names and their corresponding non-overlapping counts\n",
    "non_overlapping_counts = {\n",
    "    'CC': int(subprocess.check_output(f\"bedtools intersect -v -a {CC_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'CH': int(subprocess.check_output(f\"bedtools intersect -v -a {CH_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'HC': int(subprocess.check_output(f\"bedtools intersect -v -a {HC_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'HH': int(subprocess.check_output(f\"bedtools intersect -v -a {HH_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'allCpG': int(subprocess.check_output(f\"bedtools intersect -v -a {cpgList} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip())\n",
    "}\n",
    "\n",
    "# Create a new DataFrame with the non-overlapping counts\n",
    "non_overlapping_df = pd.DataFrame({\n",
    "    'genomicFeature': ['no_overlap'] * len(non_overlapping_counts),\n",
    "    'treatment': list(non_overlapping_counts.keys()),\n",
    "    'propOverlap': list(non_overlapping_counts.values())\n",
    "})\n",
    "\n",
    "# Append the non_overlapping_df to the original DataFrame\n",
    "df = pd.concat([df, non_overlapping_df], ignore_index=True)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f3acf60a-3c3c-4d6c-b0ae-a522e3cad965",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomicFeature</th>\n",
       "      <th>treatment</th>\n",
       "      <th>propOverlap</th>\n",
       "      <th>propOverlap_sum</th>\n",
       "      <th>proportion_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TE</td>\n",
       "      <td>CC</td>\n",
       "      <td>11289</td>\n",
       "      <td>97527</td>\n",
       "      <td>11.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exons</td>\n",
       "      <td>CC</td>\n",
       "      <td>42577</td>\n",
       "      <td>97527</td>\n",
       "      <td>43.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>introns</td>\n",
       "      <td>CC</td>\n",
       "      <td>19342</td>\n",
       "      <td>97527</td>\n",
       "      <td>19.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CC</td>\n",
       "      <td>2081</td>\n",
       "      <td>97527</td>\n",
       "      <td>2.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CC</td>\n",
       "      <td>2964</td>\n",
       "      <td>97527</td>\n",
       "      <td>3.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CC</td>\n",
       "      <td>11929</td>\n",
       "      <td>97527</td>\n",
       "      <td>12.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CC</td>\n",
       "      <td>7345</td>\n",
       "      <td>97527</td>\n",
       "      <td>7.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TE</td>\n",
       "      <td>CH</td>\n",
       "      <td>13304</td>\n",
       "      <td>112541</td>\n",
       "      <td>11.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>exons</td>\n",
       "      <td>CH</td>\n",
       "      <td>46527</td>\n",
       "      <td>112541</td>\n",
       "      <td>41.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>introns</td>\n",
       "      <td>CH</td>\n",
       "      <td>22545</td>\n",
       "      <td>112541</td>\n",
       "      <td>20.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CH</td>\n",
       "      <td>2712</td>\n",
       "      <td>112541</td>\n",
       "      <td>2.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CH</td>\n",
       "      <td>3250</td>\n",
       "      <td>112541</td>\n",
       "      <td>2.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CH</td>\n",
       "      <td>15066</td>\n",
       "      <td>112541</td>\n",
       "      <td>13.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CH</td>\n",
       "      <td>9137</td>\n",
       "      <td>112541</td>\n",
       "      <td>8.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TE</td>\n",
       "      <td>HC</td>\n",
       "      <td>19594</td>\n",
       "      <td>159867</td>\n",
       "      <td>12.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>exons</td>\n",
       "      <td>HC</td>\n",
       "      <td>57498</td>\n",
       "      <td>159867</td>\n",
       "      <td>35.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>introns</td>\n",
       "      <td>HC</td>\n",
       "      <td>34622</td>\n",
       "      <td>159867</td>\n",
       "      <td>21.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HC</td>\n",
       "      <td>4177</td>\n",
       "      <td>159867</td>\n",
       "      <td>2.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HC</td>\n",
       "      <td>4207</td>\n",
       "      <td>159867</td>\n",
       "      <td>2.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HC</td>\n",
       "      <td>24496</td>\n",
       "      <td>159867</td>\n",
       "      <td>15.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HC</td>\n",
       "      <td>15273</td>\n",
       "      <td>159867</td>\n",
       "      <td>9.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TE</td>\n",
       "      <td>HH</td>\n",
       "      <td>18394</td>\n",
       "      <td>147496</td>\n",
       "      <td>12.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>exons</td>\n",
       "      <td>HH</td>\n",
       "      <td>56100</td>\n",
       "      <td>147496</td>\n",
       "      <td>38.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>introns</td>\n",
       "      <td>HH</td>\n",
       "      <td>31070</td>\n",
       "      <td>147496</td>\n",
       "      <td>21.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HH</td>\n",
       "      <td>3704</td>\n",
       "      <td>147496</td>\n",
       "      <td>2.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HH</td>\n",
       "      <td>3986</td>\n",
       "      <td>147496</td>\n",
       "      <td>2.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HH</td>\n",
       "      <td>21279</td>\n",
       "      <td>147496</td>\n",
       "      <td>14.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HH</td>\n",
       "      <td>12963</td>\n",
       "      <td>147496</td>\n",
       "      <td>8.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TE</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2828372</td>\n",
       "      <td>23397995</td>\n",
       "      <td>12.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>exons</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2323389</td>\n",
       "      <td>23397995</td>\n",
       "      <td>9.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>introns</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>5497874</td>\n",
       "      <td>23397995</td>\n",
       "      <td>23.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>926518</td>\n",
       "      <td>23397995</td>\n",
       "      <td>3.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>600840</td>\n",
       "      <td>23397995</td>\n",
       "      <td>2.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>6644297</td>\n",
       "      <td>23397995</td>\n",
       "      <td>28.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>4576705</td>\n",
       "      <td>23397995</td>\n",
       "      <td>19.56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genomicFeature treatment  propOverlap  propOverlap_sum  \\\n",
       "0                 TE        CC        11289            97527   \n",
       "1              exons        CC        42577            97527   \n",
       "2            introns        CC        19342            97527   \n",
       "3   putativePromoter        CC         2081            97527   \n",
       "4               UTRs        CC         2964            97527   \n",
       "5         intergenic        CC        11929            97527   \n",
       "6         no_overlap        CC         7345            97527   \n",
       "7                 TE        CH        13304           112541   \n",
       "8              exons        CH        46527           112541   \n",
       "9            introns        CH        22545           112541   \n",
       "10  putativePromoter        CH         2712           112541   \n",
       "11              UTRs        CH         3250           112541   \n",
       "12        intergenic        CH        15066           112541   \n",
       "13        no_overlap        CH         9137           112541   \n",
       "14                TE        HC        19594           159867   \n",
       "15             exons        HC        57498           159867   \n",
       "16           introns        HC        34622           159867   \n",
       "17  putativePromoter        HC         4177           159867   \n",
       "18              UTRs        HC         4207           159867   \n",
       "19        intergenic        HC        24496           159867   \n",
       "20        no_overlap        HC        15273           159867   \n",
       "21                TE        HH        18394           147496   \n",
       "22             exons        HH        56100           147496   \n",
       "23           introns        HH        31070           147496   \n",
       "24  putativePromoter        HH         3704           147496   \n",
       "25              UTRs        HH         3986           147496   \n",
       "26        intergenic        HH        21279           147496   \n",
       "27        no_overlap        HH        12963           147496   \n",
       "28                TE    allCpG      2828372         23397995   \n",
       "29             exons    allCpG      2323389         23397995   \n",
       "30           introns    allCpG      5497874         23397995   \n",
       "31  putativePromoter    allCpG       926518         23397995   \n",
       "32              UTRs    allCpG       600840         23397995   \n",
       "33        intergenic    allCpG      6644297         23397995   \n",
       "34        no_overlap    allCpG      4576705         23397995   \n",
       "\n",
       "    proportion_overlap  \n",
       "0                11.58  \n",
       "1                43.66  \n",
       "2                19.83  \n",
       "3                 2.13  \n",
       "4                 3.04  \n",
       "5                12.23  \n",
       "6                 7.53  \n",
       "7                11.82  \n",
       "8                41.34  \n",
       "9                20.03  \n",
       "10                2.41  \n",
       "11                2.89  \n",
       "12               13.39  \n",
       "13                8.12  \n",
       "14               12.26  \n",
       "15               35.97  \n",
       "16               21.66  \n",
       "17                2.61  \n",
       "18                2.63  \n",
       "19               15.32  \n",
       "20                9.55  \n",
       "21               12.47  \n",
       "22               38.03  \n",
       "23               21.06  \n",
       "24                2.51  \n",
       "25                2.70  \n",
       "26               14.43  \n",
       "27                8.79  \n",
       "28               12.09  \n",
       "29                9.93  \n",
       "30               23.50  \n",
       "31                3.96  \n",
       "32                2.57  \n",
       "33               28.40  \n",
       "34               19.56  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group the data by treatment and calculate the sum of propOverlap\n",
    "treatment_sums = df.groupby('treatment')['propOverlap'].sum().reset_index()\n",
    "\n",
    "# Merge the original DataFrame with the treatment sums\n",
    "df_merged = pd.merge(df, treatment_sums, on='treatment', suffixes=('', '_sum'))\n",
    "\n",
    "# Calculate the proportion of overlap\n",
    "df_merged['proportion_overlap'] = round(df_merged['propOverlap'] / df_merged['propOverlap_sum']*100, 2)\n",
    "\n",
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "75119190-c40f-4340-9e96-9ac5303a0c18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['propOverlap'] = df['propOverlap'].str.extract('(\\d+)')\n",
    "\n",
    "df_merged.to_csv('/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/analysis/genomic_feature_tracks/majorityMe_propOverlap.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3235f7d-0dd6-474f-87ac-99982345fc75",
   "metadata": {
    "tags": []
   },
   "source": [
    "## no overlaps with genomic features\n",
    "\n",
    "CpGs that do not overlap any feature (aka unannotated intergenic regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "54199b0e-2c21-4c3c-9976-8ad7b4e40e34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4576705\n",
      "CpG motifs do not overlap with exons, introns, transposable elements (all), or putative promoters\n"
     ]
    }
   ],
   "source": [
    "# CpG motif\n",
    "!bedtools intersect -v -a {cpgList} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"CpG motifs do not overlap with exons, introns, transposable elements (all), or putative promoters\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f2206a3a-0ea8-486a-824d-9004f743101b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7345\n",
      "methylated CpG for cont cont do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n",
      "74977\n",
      "methylated CpG for cont hyp do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n",
      "101314\n",
      "methylated CpG for hyp cont do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n",
      "95457\n",
      "methylated CpG for hyp hyp do not overlap with exons, introns, transposable elements (all), intergenic regions or putative promoters\n"
     ]
    }
   ],
   "source": [
    "!bedtools intersect -v -a {CC_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"methylated CpG for cont cont do not overlap with exons, introns, transposable elements (all), or putative promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {CH_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"methylated CpG for cont hyp do not overlap with exons, introns, transposable elements (all), or putative promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {HC_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"methylated CpG for hyp cont do not overlap with exons, introns, transposable elements (all), or putative promoters\"\n",
    "\n",
    "!bedtools intersect -u -a {HH_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\n",
    "!echo \"methylated CpG for hyp hyp do not overlap with exons, introns, transposable elements (all), or putative promoters\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ef82e0b3-a29a-4c2e-b05a-2601366be7bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomicFeature</th>\n",
       "      <th>treatment</th>\n",
       "      <th>propOverlap</th>\n",
       "      <th>non_overlapping_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TE</td>\n",
       "      <td>CC</td>\n",
       "      <td>11289</td>\n",
       "      <td>7345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exons</td>\n",
       "      <td>CC</td>\n",
       "      <td>42577</td>\n",
       "      <td>7345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>introns</td>\n",
       "      <td>CC</td>\n",
       "      <td>19342</td>\n",
       "      <td>7345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CC</td>\n",
       "      <td>2081</td>\n",
       "      <td>7345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CC</td>\n",
       "      <td>2964</td>\n",
       "      <td>7345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CC</td>\n",
       "      <td>11929</td>\n",
       "      <td>7345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TE</td>\n",
       "      <td>CH</td>\n",
       "      <td>13304</td>\n",
       "      <td>74977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exons</td>\n",
       "      <td>CH</td>\n",
       "      <td>46527</td>\n",
       "      <td>74977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>introns</td>\n",
       "      <td>CH</td>\n",
       "      <td>22545</td>\n",
       "      <td>74977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>CH</td>\n",
       "      <td>2712</td>\n",
       "      <td>74977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>CH</td>\n",
       "      <td>3250</td>\n",
       "      <td>74977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>CH</td>\n",
       "      <td>15066</td>\n",
       "      <td>74977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TE</td>\n",
       "      <td>HC</td>\n",
       "      <td>19594</td>\n",
       "      <td>101314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>exons</td>\n",
       "      <td>HC</td>\n",
       "      <td>57498</td>\n",
       "      <td>101314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>introns</td>\n",
       "      <td>HC</td>\n",
       "      <td>34622</td>\n",
       "      <td>101314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HC</td>\n",
       "      <td>4177</td>\n",
       "      <td>101314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HC</td>\n",
       "      <td>4207</td>\n",
       "      <td>101314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HC</td>\n",
       "      <td>24496</td>\n",
       "      <td>101314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TE</td>\n",
       "      <td>HH</td>\n",
       "      <td>18394</td>\n",
       "      <td>95457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>exons</td>\n",
       "      <td>HH</td>\n",
       "      <td>56100</td>\n",
       "      <td>95457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>introns</td>\n",
       "      <td>HH</td>\n",
       "      <td>31070</td>\n",
       "      <td>95457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>HH</td>\n",
       "      <td>3704</td>\n",
       "      <td>95457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>HH</td>\n",
       "      <td>3986</td>\n",
       "      <td>95457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>HH</td>\n",
       "      <td>21279</td>\n",
       "      <td>95457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TE</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2828372</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>exons</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>2323389</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>introns</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>5497874</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>putativePromoter</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>926518</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>UTRs</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>600840</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>6644297</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CC</td>\n",
       "      <td>7345</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CH</td>\n",
       "      <td>74977</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HC</td>\n",
       "      <td>101314</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HH</td>\n",
       "      <td>95457</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>9881998</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CC</td>\n",
       "      <td>7345</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>CH</td>\n",
       "      <td>9137</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HC</td>\n",
       "      <td>15273</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>HH</td>\n",
       "      <td>12963</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>no_overlap</td>\n",
       "      <td>allCpG</td>\n",
       "      <td>4576705</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genomicFeature treatment  propOverlap non_overlapping_count\n",
       "0                 TE        CC        11289                  7345\n",
       "1              exons        CC        42577                  7345\n",
       "2            introns        CC        19342                  7345\n",
       "3   putativePromoter        CC         2081                  7345\n",
       "4               UTRs        CC         2964                  7345\n",
       "5         intergenic        CC        11929                  7345\n",
       "6                 TE        CH        13304                 74977\n",
       "7              exons        CH        46527                 74977\n",
       "8            introns        CH        22545                 74977\n",
       "9   putativePromoter        CH         2712                 74977\n",
       "10              UTRs        CH         3250                 74977\n",
       "11        intergenic        CH        15066                 74977\n",
       "12                TE        HC        19594                101314\n",
       "13             exons        HC        57498                101314\n",
       "14           introns        HC        34622                101314\n",
       "15  putativePromoter        HC         4177                101314\n",
       "16              UTRs        HC         4207                101314\n",
       "17        intergenic        HC        24496                101314\n",
       "18                TE        HH        18394                 95457\n",
       "19             exons        HH        56100                 95457\n",
       "20           introns        HH        31070                 95457\n",
       "21  putativePromoter        HH         3704                 95457\n",
       "22              UTRs        HH         3986                 95457\n",
       "23        intergenic        HH        21279                 95457\n",
       "24                TE    allCpG      2828372                   NaN\n",
       "25             exons    allCpG      2323389                   NaN\n",
       "26           introns    allCpG      5497874                   NaN\n",
       "27  putativePromoter    allCpG       926518                   NaN\n",
       "28              UTRs    allCpG       600840                   NaN\n",
       "29        intergenic    allCpG      6644297                   NaN\n",
       "30        no_overlap        CC         7345                   NaN\n",
       "31        no_overlap        CH        74977                   NaN\n",
       "32        no_overlap        HC       101314                   NaN\n",
       "33        no_overlap        HH        95457                   NaN\n",
       "34        no_overlap    allCpG      9881998                   NaN\n",
       "35        no_overlap        CC         7345                   NaN\n",
       "36        no_overlap        CH         9137                   NaN\n",
       "37        no_overlap        HC        15273                   NaN\n",
       "38        no_overlap        HH        12963                   NaN\n",
       "39        no_overlap    allCpG      4576705                   NaN"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "# Create a dictionary with treatment names and their corresponding non-overlapping counts\n",
    "non_overlapping_counts = {\n",
    "    'CC': int(subprocess.check_output(f\"bedtools intersect -v -a {CC_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'CH': int(subprocess.check_output(f\"bedtools intersect -v -a {CH_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'HC': int(subprocess.check_output(f\"bedtools intersect -v -a {HC_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'HH': int(subprocess.check_output(f\"bedtools intersect -v -a {HH_CpG} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip()),\n",
    "    'allCpG': int(subprocess.check_output(f\"bedtools intersect -v -a {cpgList} -b {exonList} {intronList} {teList} {promoterList} | wc -l\", shell=True).decode().strip())\n",
    "}\n",
    "\n",
    "# Create a new DataFrame with the non-overlapping counts\n",
    "non_overlapping_df = pd.DataFrame({\n",
    "    'genomicFeature': ['no_overlap'] * len(non_overlapping_counts),\n",
    "    'treatment': list(non_overlapping_counts.keys()),\n",
    "    'propOverlap': list(non_overlapping_counts.values())\n",
    "})\n",
    "\n",
    "# Append the non_overlapping_df to the original DataFrame\n",
    "df = pd.concat([df, non_overlapping_df], ignore_index=True)\n",
    "\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-bedtools]",
   "language": "python",
   "name": "conda-env-.conda-bedtools-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
