{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b9fbce3a-245f-41eb-8233-9d7df692d304",
   "metadata": {},
   "source": [
    "bash script for initial fastqc on raw reads - important to specify an output directory (with -o flag)\n",
    "\n",
    "sarah said any job under 24 hours is still considered a short job, so good to just have that be your default so things don't run out of time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c39740-ad02-4a1e-a02f-09311eaba317",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "module load miniconda/4.11.0\n",
    "\n",
    "conda config --add channels defaults\n",
    "conda config --add channels bioconda\n",
    "conda config --add channels conda-forge\n",
    "conda config --set channel_priority strict\n",
    "\n",
    "conda create --name myenv python=3.11\n",
    "conda activate myenv\n",
    "\n",
    "\n",
    "conda install fastqc\n",
    "\n",
    "\n",
    "fastqc -o /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/fastqc /project/pi_sarah_gignouxwolfsohn_uml_edu/Raw_sequences/methyl_raw/*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dea4895c-4d8c-4f29-b504-e43509b9e76d",
   "metadata": {},
   "source": [
    "easier to look and interpret the fastqc results with multiqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc7a0369-a2d1-4ab4-a387-d042029b43d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "module load miniconda/4.11.0\n",
    "\n",
    "conda config --add channels defaults\n",
    "conda config --add channels bioconda\n",
    "conda config --add channels conda-forge\n",
    "conda config --set channel_priority strict\n",
    "\n",
    "conda create --name myenv python=3.11\n",
    "conda activate myenv\n",
    "\n",
    "conda install multiqc\n",
    "\n",
    "\n",
    "multiqc -o /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/multiqc /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/fastqc/fastqc_html/*\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3013707b-cd4c-4b82-a8cd-8f416c5e993e",
   "metadata": {},
   "source": [
    "results from this is an html link - have to download it, then open it on a browser\n",
    "\n",
    "assuming everything looks okay (or good enough, quality won't be there yet) - trim reads - I used trim-galore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10769de9-3959-40aa-87c2-10e8440bcf8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "#-----------------modules-----------------#\n",
    "module load miniconda/4.11.0\n",
    "\n",
    "# have to make sure you have cutadapt AND fastqc installed before trim_galore\n",
    "\n",
    "conda create -n cutadaptenv cutadapt\n",
    "conda activate cutadaptenv\n",
    "\n",
    "conda install trim-galore\n",
    "\n",
    "#---------------change wd----------------#\n",
    "\n",
    "cd project/pi_sarah_gignouxwolfsohn_uml_edu/Raw_sequences/methyl_raw/\n",
    "\n",
    "#-----------------commands----------------#\n",
    "\n",
    "for f in $(ls project/pi_sarah_gignouxwolfsohn_uml_edu/Raw_sequences/methyl_raw/ | sed 's/[^-]*$//'); \n",
    "do \n",
    "    trim_galore -q 20 --phred33 --length 20 --max_length 40 --paired \"$f\"CV_R1_001.fastq.gz \"$f\"CV_R2_001.fastq.gz --output_dir project/pi_sarah_gignouxwolfsohn_uml_edu/julia/trimmed/filtered_auto_trim_sequences/;\n",
    "done\n",
    "\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49e714d4-b15f-4dec-ac3c-95e8bff9ea86",
   "metadata": {},
   "source": [
    "phred score min of 33, min length 20bp, max length 40bp, paired-ends, didn't specify an adapter to remove so the default is auto-detect - and need to specify output directory\n",
    "\n",
    "then fastqc and multiqc on these filtered and trimmed samples and compared - made sure quality checks were better\n",
    "\n",
    "for mapping, tried hisat2 but was getting low alignment rates - sticking with bowtie2 instead - get alignment rates between ~85-90%\n",
    "\n",
    "Submitted GenBank assembly GCA_002022765.4 https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_002022765.2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6087c50-6eac-459a-9299-cc05077b8484",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "# ----------------Modules------------------------- #\n",
    "\n",
    "module load miniconda/4.11.0\n",
    "module load bowtie2/2.4.2\n",
    "\n",
    "#------------build index------------------#\n",
    "\n",
    "# This line creates the index in the current directory\n",
    "bowtie2-build GCA_002022765.4_C_virginica-3.0_genomic.fna reference_index\n",
    "\n",
    "# Set the paths and directories\n",
    "reference_index=\"reference_index\"\n",
    "\n",
    "# Directory containing your input FASTQ files\n",
    "input_dir=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/trimmed/filtered_auto_trim_sequences/trim_files\"\n",
    "\n",
    "# Output directory for Bowtie2 results\n",
    "output_dir=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2/filtered_alignment_output\"\n",
    "\n",
    "# Create an array with your sample names\n",
    "sample_names=( \"2018--BBB-WBO-B21-CV\"\n",
    "\"2018--BBB-WBV-B70-CV\"\n",
    "\"2018--BBO-BBO-B16-CV\"\n",
    "\"2018--BBO-BBY-B27-CV\"\n",
    "\"2018--BBO-WBO-B16-CV\"\n",
    "\"2018--BBO-WBV-B64-CV\"\n",
    "\"2018--BBR-BBB-B50-CV\"\n",
    "\"2018--BBR-BBG-B38-CV\"\n",
    "\"2018--BBR-BBY-B26-CV\"\n",
    "\"2018--BBY-WBG-B42-CV\"\n",
    "\"2018--BPO-BPO-O16-CV\"\n",
    "\"2018--BPR-BPG-O38-CV\"\n",
    "\"2018--BPR-BPR-O02-CV\"\n",
    "\"2018--BPY-BPG-O42-CV\"\n",
    "\"2018--BPY-BPY-O29-CV\"\n",
    "\"2018--WBB-WBV-W69-CV\"\n",
    "\"2018--WBG-BBB-W56-CV\"\n",
    "\"2018--WBG-WBG-W44-CV\"\n",
    "\"2018--WBO-BBR-W03-CV\"\n",
    "\"2018--WBO-WBV-W64-CV\"\n",
    "\"2018--WBR-BBY-W25-CV\"\n",
    "\"2018--WBV-WBO-W23-CV\"\n",
    "\"2018--WBV-WBR-W12-CV\"\n",
    "\"2018--WBY-BBV-W65-CV\"\n",
    "\"2018--WBY-BBY-W30-CV\"\n",
    "\"2018--WPB-BPG-G45-CV\"\n",
    "\"2018--WPO-BPO-G16-CV\"\n",
    "\"2018--WPO-BPY-G28-CV\"\n",
    "\"2018--WPR-BPY-G25-CV\"\n",
    "\"2018--WPV-BPR-G11-CV\" )\n",
    "\n",
    "\n",
    "# Loop through the sample names\n",
    "for sample_name in \"${sample_names[@]}\"; do\n",
    "    # Construct the file names for R1 and R2\n",
    "    read1=\"${input_dir}/${sample_name}_R1_001_val_1.fq.gz\"\n",
    "    read2=\"${input_dir}/${sample_name}_R2_001_val_2.fq.gz\"\n",
    "\n",
    "    # Output SAM file with full path to the output directory\n",
    "    output_sam=\"${output_dir}/${sample_name}_alignment.sam\"\n",
    "\n",
    "    # Run Bowtie2 for paired-end reads\n",
    "    bowtie2 --very-sensitive --local -x \"${reference_index}\" -1 \"${read1}\" -2 \"${read2}\" -S \"${output_sam}\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f97d23f7-a88f-4d59-9303-b1402cec2283",
   "metadata": {},
   "source": [
    "need to find another way to do the sample names better - more automated - but struggling with getting that part of the code to work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb7ec910-cc7c-497b-a51a-8a6182304e13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the list to store the matching lines\n",
    "alignment_rate_lines = []\n",
    "\n",
    "# Specify the Slurm output file path\n",
    "slurm_out_file = '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2/slurm-13167043.out'\n",
    "\n",
    "# Open the Slurm output file and search for lines containing 'overall alignment rate'\n",
    "with open(slurm_out_file, 'r') as file:\n",
    "    for line in file:\n",
    "        if 'overall alignment rate' in line:\n",
    "            alignment_rate_lines.append(line.strip())\n",
    "\n",
    "# Print the extracted lines or perform further processing\n",
    "for line in alignment_rate_lines:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4fe53b0-6b4d-4ebb-b651-1144d8304009",
   "metadata": {},
   "source": [
    "finding the overall alignment rates for each sample from the slurm.out file\n",
    "\n",
    "88.69% overall alignment rate\n",
    "87.89% overall alignment rate\n",
    "88.79% overall alignment rate\n",
    "88.51% overall alignment rate\n",
    "88.17% overall alignment rate\n",
    "88.19% overall alignment rate\n",
    "88.64% overall alignment rate\n",
    "88.59% overall alignment rate\n",
    "88.58% overall alignment rate\n",
    "89.15% overall alignment rate\n",
    "88.38% overall alignment rate\n",
    "88.77% overall alignment rate\n",
    "88.71% overall alignment rate\n",
    "88.28% overall alignment rate\n",
    "88.79% overall alignment rate\n",
    "87.70% overall alignment rate\n",
    "88.44% overall alignment rate\n",
    "88.46% overall alignment rate\n",
    "87.39% overall alignment rate\n",
    "87.72% overall alignment rate\n",
    "88.25% overall alignment rate\n",
    "88.66% overall alignment rate\n",
    "88.38% overall alignment rate\n",
    "88.12% overall alignment rate\n",
    "88.34% overall alignment rate\n",
    "87.25% overall alignment rate\n",
    "88.63% overall alignment rate\n",
    "87.33% overall alignment rate\n",
    "88.85% overall alignment rate\n",
    "88.54% overall alignment rate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "435c5686-2c23-4887-84ec-88691867d280",
   "metadata": {},
   "source": [
    "now need to convert SAM files to BAM files and sort those for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "123d010c-bb49-4d8f-8abd-f20fb3ded067",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "# ----------------Modules------------------------- #\n",
    "\n",
    "module load miniconda/4.11.0\n",
    "module load samtools/1.9\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME\n",
    "#\n",
    "\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2/SAM_files/* | sed 's/[^-]*$//'); \n",
    "do \n",
    "samtools view -b \"$f\"CV_alignment.sam > \"$f\"CV_alignment.bam;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a533b08-b0ed-495e-b4aa-53abd7593902",
   "metadata": {},
   "source": [
    "now have BAM files, and needed to move the BAM files to the right directory - now sorting BAM files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "281c5464-5cd3-4ed3-baf7-25fb87124097",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "# ----------------Modules------------------------- #\n",
    "#\n",
    "module load miniconda/4.11.0\n",
    "module load samtools/1.9\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME\n",
    "#\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2/BAM_files/* | sed 's/[^-]*$//'); \n",
    "do \n",
    "samtools sort \"$f\"CV_alignment.bam -o \"$f\"CV_sorted.bam;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22fa7085-9b1a-4bf6-9f6c-5aeebf90031b",
   "metadata": {},
   "source": [
    "now have sorted BAM files - need to run picard tool and mark optical duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8026bdf7-077d-4506-a066-3a78dda9ba6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "# ----------------Parameters---------------------- #\n",
    "#\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "#\n",
    "# ----------------Modules------------------------- #\n",
    "#\n",
    "module load miniconda/4.11.0\n",
    "\n",
    "conda create -n my-java-environment -c conda-forge openjdk=17\n",
    "conda activate my-java-environment\n",
    "\n",
    "\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "# Define the path to the Picard JAR file\n",
    "PICARD_JAR=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/picard.jar\"  # Replace with the actual path\n",
    "\n",
    "echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME\n",
    "#\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2/sorted_BAM_files/* | sed 's/[^-]*$//'); \n",
    "do \n",
    "java -jar \"$PICARD_JAR\" MarkDuplicates \\\n",
    "      I=\"$f\"CV_sorted.bam \\\n",
    "      O=\"$f\"marked_duplicates.bam \\\n",
    "      M=\"$f\"marked_dup_metrics.txt;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "980f1a61-4aed-4c28-abbc-dd366f742ce4",
   "metadata": {},
   "source": [
    "so i checked sarah's code - she made notes that she should've aligned to the RefSeq assembly since it is annotated - so I'm going to go back through the assembly and picard duplicate steps with the RefSeq file\n",
    "\n",
    "NCBI RefSeq assembly: GCF_002022765.2 https://www.ncbi.nlm.nih.gov/genome/annotation_euk/Crassostrea_virginica/100/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acd90a30-5025-4e9a-bb3b-86d0a6b0753c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "# ----------------Modules------------------------- #\n",
    "\n",
    "module load miniconda/4.11.0\n",
    "module load bowtie2/2.4.2\n",
    "\n",
    "#------------build index------------------#\n",
    "\n",
    "# This line creates the index in the current directory\n",
    "bowtie2-build GCF_002022765.2_C_virginica-3.0_genomic.fna reference_index\n",
    "\n",
    "# Set the paths and directories\n",
    "reference_index=\"reference_index\"\n",
    "\n",
    "# Directory containing your input FASTQ files\n",
    "input_dir=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/trimmed/filtered_auto_trim_sequences/trim_files\"\n",
    "\n",
    "# Output directory for Bowtie2 results\n",
    "output_dir=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2/filtered_alignment_output\"\n",
    "\n",
    "# Create an array with your sample names\n",
    "sample_names=( \"2018--BBB-WBO-B21-CV\"\n",
    "\"2018--BBB-WBV-B70-CV\"\n",
    "\"2018--BBO-BBO-B16-CV\"\n",
    "\"2018--BBO-BBY-B27-CV\"\n",
    "\"2018--BBO-WBO-B16-CV\"\n",
    "\"2018--BBO-WBV-B64-CV\"\n",
    "\"2018--BBR-BBB-B50-CV\"\n",
    "\"2018--BBR-BBG-B38-CV\"\n",
    "\"2018--BBR-BBY-B26-CV\"\n",
    "\"2018--BBY-WBG-B42-CV\"\n",
    "\"2018--BPO-BPO-O16-CV\"\n",
    "\"2018--BPR-BPG-O38-CV\"\n",
    "\"2018--BPR-BPR-O02-CV\"\n",
    "\"2018--BPY-BPG-O42-CV\"\n",
    "\"2018--BPY-BPY-O29-CV\"\n",
    "\"2018--WBB-WBV-W69-CV\"\n",
    "\"2018--WBG-BBB-W56-CV\"\n",
    "\"2018--WBG-WBG-W44-CV\"\n",
    "\"2018--WBO-BBR-W03-CV\"\n",
    "\"2018--WBO-WBV-W64-CV\"\n",
    "\"2018--WBR-BBY-W25-CV\"\n",
    "\"2018--WBV-WBO-W23-CV\"\n",
    "\"2018--WBV-WBR-W12-CV\"\n",
    "\"2018--WBY-BBV-W65-CV\"\n",
    "\"2018--WBY-BBY-W30-CV\"\n",
    "\"2018--WPB-BPG-G45-CV\"\n",
    "\"2018--WPO-BPO-G16-CV\"\n",
    "\"2018--WPO-BPY-G28-CV\"\n",
    "\"2018--WPR-BPY-G25-CV\"\n",
    "\"2018--WPV-BPR-G11-CV\" )\n",
    "\n",
    "\n",
    "# Loop through the sample names\n",
    "for sample_name in \"${sample_names[@]}\"; do\n",
    "    # Construct the file names for R1 and R2\n",
    "    read1=\"${input_dir}/${sample_name}_R1_001_val_1.fq.gz\"\n",
    "    read2=\"${input_dir}/${sample_name}_R2_001_val_2.fq.gz\"\n",
    "\n",
    "    # Output SAM file with full path to the output directory\n",
    "    output_sam=\"${output_dir}/${sample_name}_alignment.sam\"\n",
    "\n",
    "    # Run Bowtie2 for paired-end reads\n",
    "    bowtie2 --very-sensitive --local -x \"${reference_index}\" -1 \"${read1}\" -2 \"${read2}\" -S \"${output_sam}\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "695dd3f7-b91d-4e42-90a2-259edfa4bb7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88.69% overall alignment rate\n",
      "87.89% overall alignment rate\n",
      "88.79% overall alignment rate\n",
      "88.51% overall alignment rate\n",
      "88.17% overall alignment rate\n",
      "88.19% overall alignment rate\n",
      "88.64% overall alignment rate\n",
      "88.59% overall alignment rate\n",
      "88.58% overall alignment rate\n",
      "89.15% overall alignment rate\n",
      "88.38% overall alignment rate\n",
      "88.77% overall alignment rate\n",
      "88.71% overall alignment rate\n",
      "88.29% overall alignment rate\n",
      "88.79% overall alignment rate\n",
      "87.70% overall alignment rate\n",
      "88.44% overall alignment rate\n",
      "88.46% overall alignment rate\n",
      "87.39% overall alignment rate\n",
      "87.72% overall alignment rate\n",
      "88.25% overall alignment rate\n",
      "88.66% overall alignment rate\n",
      "88.38% overall alignment rate\n",
      "88.12% overall alignment rate\n",
      "88.34% overall alignment rate\n",
      "87.25% overall alignment rate\n",
      "88.63% overall alignment rate\n",
      "87.33% overall alignment rate\n",
      "88.85% overall alignment rate\n",
      "88.54% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "# Define the list to store the matching lines\n",
    "alignment_rate_lines = []\n",
    "\n",
    "# Specify the Slurm output file path\n",
    "slurm_out_file = '/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2_refseq/slurm-13289020.out'\n",
    "\n",
    "# Open the Slurm output file and search for lines containing 'overall alignment rate'\n",
    "with open(slurm_out_file, 'r') as file:\n",
    "    for line in file:\n",
    "        if 'overall alignment rate' in line:\n",
    "            alignment_rate_lines.append(line.strip())\n",
    "\n",
    "# Print the extracted lines or perform further processing\n",
    "for line in alignment_rate_lines:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f6c064a-ce3f-4bcb-9d0a-cbf693b1b001",
   "metadata": {},
   "source": [
    "overall alignment rates:\n",
    "88.69% overall alignment rate 87.89% overall alignment rate 88.79% overall alignment rate 88.51% overall alignment rate 88.17% overall alignment rate 88.19% overall alignment rate 88.64% overall alignment rate 88.59% overall alignment rate 88.58% overall alignment rate 89.15% overall alignment rate 88.38% overall alignment rate 88.77% overall alignment rate 88.71% overall alignment rate 88.29% overall alignment rate 88.79% overall alignment rate 87.70% overall alignment rate 88.44% overall alignment rate 88.46% overall alignment rate 87.39% overall alignment rate 87.72% overall alignment rate 88.25% overall alignment rate 88.66% overall alignment rate 88.38% overall alignment rate 88.12% overall alignment rate 88.34% overall alignment rate 87.25% overall alignment rate 88.63% overall alignment rate 87.33% overall alignment rate 88.85% overall alignment rate 88.54% overall alignment rate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecd18516-3aef-4774-85cb-8b3600a25956",
   "metadata": {},
   "source": [
    "converting SAM to BAM files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37ad254a-5ee6-441a-9fa0-983fda517171",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "# ----------------Modules------------------------- #\n",
    "\n",
    "module load miniconda/4.11.0\n",
    "module load samtools/1.9\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME\n",
    "#\n",
    "\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2_refseq/SAM_files/* | sed 's/[^-]*$//'); \n",
    "do \n",
    "samtools view -b \"$f\"CV_alignment.sam > \"$f\"CV_alignment.bam;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ab9c434-9b64-425d-8266-6fc5aea01657",
   "metadata": {},
   "source": [
    "sorting BAM files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86d41493-951b-4cbe-8d25-ec7d81f3a325",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "\n",
    "# ----------------Modules------------------------- #\n",
    "#\n",
    "module load miniconda/4.11.0\n",
    "module load samtools/1.9\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME\n",
    "#\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2_refseq/BAM_files/* | sed 's/[^-]*$//'); \n",
    "do \n",
    "samtools sort \"$f\"CV_alignment.bam -o \"$f\"CV_sorted.bam;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff7ddc5f-afc3-40a2-90c7-3a5e3b6039b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "now using picard to mark duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05674b7c-95c0-4ac0-87f8-3be10bbd7325",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "# ----------------Parameters---------------------- #\n",
    "#\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "#\n",
    "# ----------------Modules------------------------- #\n",
    "#\n",
    "module load miniconda/4.11.0\n",
    "#\n",
    "conda create -n my-java-environment -c conda-forge openjdk=17\n",
    "conda activate my-java-environment\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "# Define the path to the Picard JAR file\n",
    "PICARD_JAR=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/julia/CE_MethylRAD_analysis_2018/picard.jar\"  # Replace with the actual path\n",
    "\n",
    "echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME\n",
    "#\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2_refseq/sorted_BAM_files/* | sed 's/[^-]*$//'); \n",
    "do \n",
    "java -jar \"$PICARD_JAR\" MarkDuplicates \\\n",
    "      I=\"$f\"CV_sorted.bam \\\n",
    "      O=\"$f\"marked_duplicates.bam \\\n",
    "      M=\"$f\"marked_dup_metrics.txt;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24bcc20e-1274-4ec1-adf9-928493aa5711",
   "metadata": {},
   "source": [
    "to use bedtools multicov (reports the count of alignments from multiple position-sorted and indexed BAM files that overlap intervals in a BED file)\n",
    "\n",
    "first need to index BAM files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cee7b78f-24e3-4762-ae0c-0cc52d3f5bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 1:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "#\n",
    "# ----------------Modules------------------------- #\n",
    "#\n",
    "module load miniconda/4.11.0\n",
    "module load samtools/1.9\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "#\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2_refseq/sorted_BAM_files/*_sorted.bam); \n",
    "do \n",
    "samtools index -b \"$f\" \"$f\".bai;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7379e91d-557b-45da-92ee-ea230a89efaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# converting gff to bed file and sorting with bedtools\n",
    "bedtools sort -i genomic.gff > GCF_002022765.2_C_virginica-3.0_genomic.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d0a4382-afbc-4866-9335-60a9589e129d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "#\n",
    "# ----------------Modules------------------------- #\n",
    "module load bedtools/2\n",
    "#\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "bedtools multicov -D -bams /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bedtools/bai_index/*.bai -bed GCF_002022765.2_C_virginica-3.0_genomic.bed\n",
    "#\n",
    "echo = `date` job $JOB_NAME done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d65b4522-7c44-491b-9507-297abddab7ce",
   "metadata": {},
   "source": [
    "trying htseq-counts since I'm having trouble with bedtools multicov\n",
    "although, i used the gtf file instead of the gff file, so maybe that's the problem with the code above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f92d65a-e165-49f0-9122-5bd52a9df492",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "#\n",
    "# ----------------Modules------------------------- #\n",
    "pip install HTseq\n",
    "\n",
    "# ----------------Your Commands------------------- #\n",
    "#\n",
    "#\n",
    "for f in $(ls /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2_refseq/sorted_BAM_files/*_sorted.bam); \n",
    "do \n",
    "htseq-count -f bam \"$f\" /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/reference_genomes/genomic.gtf;\n",
    "done\n",
    "#\n",
    "echo = `date` job $JOB_NAME done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce948df4-614e-40f9-ad40-9c122e09e3f5",
   "metadata": {},
   "source": [
    "I actually think this is the better way to do it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a600b3e3-b9ef-44c9-bc50-fed800dc9a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 4  # Number of Cores per Task\n",
    "#SBATCH --mem=8192  # Requested Memory\n",
    "#SBATCH -p cpu-long  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o slurm-%j.out  # %j = job ID\n",
    "#\n",
    "# ----------------Modules------------------------- #\n",
    "pip install HTseq\n",
    "# ----------------Your Commands------------------- #\n",
    "\n",
    "htseq-count -r pos -f bam /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/bowtie2_refseq/sorted_BAM_files/*_sorted.bam /project/pi_sarah_gignouxwolfsohn_uml_edu/julia/assembly/reference_genomes/genomic.gtf >counts.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
